{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp Runner"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Runner module\n",
    "\n",
    "> API details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "from __future__ import print_function\n",
    "import os\n",
    "from SEQLinkage.Utils import *\n",
    "from collections import deque, defaultdict\n",
    "from os.path import splitext, basename, isdir, isfile\n",
    "from itertools import chain\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import brewer2mpl\n",
    "from scipy.optimize import minimize_scalar\n",
    "import re, glob\n",
    "from shutil import rmtree as remove_tree\n",
    "from distutils.file_util import copy_file\n",
    "#formatters\n",
    "#the handler, called from main, can call specific formatter.\n",
    "def format(tpeds, tfam, prev = None, wild_pen = None, muta_pen = None, out_format = 'MERLIN', inherit_mode = None, theta_max = None, theta_inc = None):\n",
    "    print(env.jobs) #testing line\n",
    "    if out_format == 'plink':\n",
    "        parmap(lambda x: format_plink(x, tfam), tpeds, env.jobs)\n",
    "    elif out_format == 'mega2':\n",
    "        mkpath(os.path.join(env.outdir, 'MEGA2'))\n",
    "        parmap(lambda x: format_mega2(x, tfam), tpeds, env.jobs)\n",
    "    elif out_format == 'merlin':\n",
    "        mkpath(os.path.join(env.outdir, 'MERLIN'))\n",
    "        parmap(lambda x: format_merlin(x, tfam), tpeds, env.jobs)\n",
    "    elif out_format == 'linkage':\n",
    "        parmap(lambda x: format_linkage(x, tfam, prev, wild_pen, muta_pen, inherit_mode, theta_max, theta_inc), tpeds, env.jobs)\n",
    "\n",
    "#plink format, ped and map\n",
    "def format_plink(tped, tfam):\n",
    "    out_base = '{}/PLINK/{}'.format(env.outdir, splitext(basename(tped))[0])\n",
    "    with open(tped) as tped_fh, open(tfam) as tfam_fh:\n",
    "        geno = []\n",
    "        with open(out_base + '.map', 'w') as m:\n",
    "            for line in tped_fh:\n",
    "                s = line.strip().split()\n",
    "                m.write(env.delimiter.join(s[:3]) + '\\n')\n",
    "                geno.append(deque(s[4:]))\n",
    "        with open(out_base + '.ped', 'w') as p:\n",
    "            for line in tfam_fh:\n",
    "                p.write(line.strip())\n",
    "                list(map(lambda x: p.write(' {} {}'.format(x.popleft(), x.popleft)), geno))\n",
    "                p.write(\"\\n\")\n",
    "\n",
    "#mega2 format, datain.01, pedin.01, map.01\n",
    "def format_mega2(tped, tfam):\n",
    "    trait = 'A' if env.trait == 'binary' else 'T'\n",
    "    pedheader = ['Pedigree', 'ID', 'Father', 'Mother', 'Sex', 'Trait.{}'.format(trait)]\n",
    "    out_base = os.path.join(env.outdir, 'MEGA2')\n",
    "    suffix =  re.search(r'chr([0-9XY]+)', basename(tped)).groups()[0]\n",
    "    with open(tped) as tph, open(tfam) as tfh:\n",
    "        geno = []\n",
    "        name = []\n",
    "        with open('{}/map.{}'.format(out_base, suffix), 'w') as m,\\\n",
    "             open('{}/datain.{}'.format(out_base, suffix), 'w') as d:\n",
    "            d.write('Type\\tName\\n{}\\tTrait\\n'.format(trait))\n",
    "            m.write('Chromosome\\tMap.k.a\\tName\\tMap.k.m\\tMap.k.f\\thg19.p\\n')\n",
    "            for line in tph:\n",
    "                s = line.strip().split()\n",
    "                gene, gno = re.search(r'^(\\S+?)(?:\\[(\\d+)\\])?$', s[1]).groups()\n",
    "                if not gno:\n",
    "                    gno = '0'\n",
    "                    with env.format_counter.get_lock():\n",
    "                        env.format_counter.value += 1\n",
    "                elif gno == '1':\n",
    "                    with env.format_counter.get_lock():\n",
    "                        env.format_counter.value += 1\n",
    "                if env.format_counter.value % (env.batch * env.jobs) == 0:\n",
    "                    env.log('{:,d} units processed {{{:.2%}}} ...'.format(env.format_counter.value, float(env.format_counter.value)/env.success_counter.value), flush=True)\n",
    "                d.write('M\\t{}\\n'.format(s[1]))\n",
    "                dis = s[2].split(';')\n",
    "                dis.insert(1,s[1])\n",
    "                m.write('{}\\t{}\\t{}\\n'.format(s[0],'\\t'.join(dis),s[3]))\n",
    "                name.append('{0}.M.1\\t{0}.M.2'.format(s[1]))\n",
    "                geno.append(deque(map(lambda x: re.sub(r'^0$', 'NA', x), s[4:])))\n",
    "        with open('{}/pedin.{}'.format(out_base, suffix), 'w') as p:\n",
    "            p.write('{}\\n'.format('\\t'.join(pedheader + name)))\n",
    "            for line in tfh:\n",
    "                p.write(line.strip())\n",
    "                s = line.strip().split()\n",
    "                list(map(lambda x: p.write('\\t{}\\t{}'.format(x.popleft(), x.popleft())), geno))\n",
    "                p.write(\"\\n\")\n",
    "\n",
    "#merlin format\n",
    "def format_merlin(tped, tfam):\n",
    "    trait = 'A' if env.trait == 'binary' else 'T'\n",
    "    #pedheader = ['Pedigree', 'ID', 'Father', 'Mother', 'Sex', 'Trait.{}'.format(trait)]\n",
    "    out_base = os.path.join(env.outdir, 'MERLIN', splitext(basename(tped))[0])\n",
    "    #suffix =  re.search(r'chr([0-9XY]+)', basename(tped)).groups()[0]\n",
    "    with open(tped) as tph, open(tfam) as tfh:\n",
    "        geno = []\n",
    "        #name = []\n",
    "        with open('{}.map'.format(out_base), 'w') as m,\\\n",
    "             open('{}.dat'.format(out_base), 'w') as d:\n",
    "            d.write('{}\\tTrait\\n'.format(trait))\n",
    "            m.write('CHROMOSOME\\tMARKER\\t\\tPOSITION\\tFEMALE_POSITION\\tMALE_POSITION\\n')\n",
    "            for line in tph:\n",
    "                s = line.strip().split()\n",
    "                gene, gno = re.search(r'^(\\S+?)(?:\\[(\\d+)\\])?$', s[1]).groups()\n",
    "                if not gno:\n",
    "                    gno = '0'\n",
    "                    with env.format_counter.get_lock():\n",
    "                        env.format_counter.value += 1\n",
    "                elif gno == '1':\n",
    "                    with env.format_counter.get_lock():\n",
    "                        env.format_counter.value += 1\n",
    "                if env.format_counter.value % (env.batch * env.jobs) == 0:\n",
    "                    env.log('{:,d} units processed {{{:.2%}}} ...'.format(env.format_counter.value, float(env.format_counter.value)/env.success_counter.value), flush=True)\n",
    "                d.write('M\\t{}\\n'.format(s[1]))\n",
    "                dis = s[2].split(';')\n",
    "                dis[1], dis[2] = dis[2], dis[1]\n",
    "                #dis.insert(1,s[1])\n",
    "                m.write('{}\\t{}\\t{}\\n'.format(s[0],s[1],'\\t'.join(dis)))\n",
    "                #name.append('{0}.M.1\\t{0}.M.2'.format(s[1]))\n",
    "                geno.append(deque(s[4:]))\n",
    "        with open('{}.ped'.format(out_base), 'w') as p:\n",
    "            #p.write('{}\\n'.format('\\t'.join(pedheader + name)))\n",
    "            for line in tfh:\n",
    "                p.write(line.strip())\n",
    "                s = line.strip().split()\n",
    "                list(map(lambda x: p.write('\\t{}\\t{}'.format(x.popleft(), x.popleft())), geno))\n",
    "                p.write(\"\\n\")\n",
    "\n",
    "\n",
    "#linkage format, .pre and .loc\n",
    "#per locus, per family based\n",
    "#because the haplotype patterns are different from family to family.\n",
    "#You can analyze them all together\n",
    "def format_linkage(tped, tfam, prev, wild_pen, muta_pen, inherit_mode, theta_max, theta_inc):\n",
    "    out_base = '{}/LINKAGE/{}'.format(env.tmp_dir, splitext(basename(tped))[0])\n",
    "    with open(tped) as tped_fh, open(tfam) as tfam_fh:\n",
    "        fams = parse_tfam(tfam_fh)\n",
    "        #parse per family per locus AF file\n",
    "        af = defaultdict(lambda: [])\n",
    "        #try to open the file for allele frequencies, otherwise use the default value\n",
    "        try:\n",
    "            with open(os.path.join(env.tmp_cache, basename(out_base) + '.freq')) as af_fh:\n",
    "                for line in af_fh:\n",
    "                    s = line.strip().split()\n",
    "                    freq = list(map(lambda x: max(1e-3, float(x)), s[2:]))\n",
    "                    relativefreq = np.array(freq)/sum(freq)\n",
    "                    af[(s[0],s[1])] = list(map(str, relativefreq))\n",
    "        except IOError:\n",
    "            env.error('freq info not properly read for [{}]'.format(basename(out_base)))\n",
    "        #parse tped\n",
    "        heter_pen = wild_pen\n",
    "        if inherit_mode == 'AD':\n",
    "            heter_pen = muta_pen\n",
    "        for line in tped_fh:\n",
    "            s = line.strip().split()\n",
    "            gene, gno = re.search(r'^(\\S+?)(?:\\[(\\d+)\\])?$', s[1]).groups()\n",
    "            if not gno:\n",
    "                gno = '0'\n",
    "                with env.format_counter.get_lock():\n",
    "                    env.format_counter.value += 1\n",
    "            elif gno == '1':\n",
    "                with env.format_counter.get_lock():\n",
    "                    env.format_counter.value += 1\n",
    "            if env.format_counter.value % (env.batch * env.jobs) == 0:\n",
    "                env.log('{:,d} units processed {{{:.2%}}} ...'.format(env.format_counter.value, float(env.format_counter.value)/env.success_counter.value), flush=True)\n",
    "            for fid in fams:\n",
    "                workdir = '{}/{}/{}'.format(out_base, gene, fid)\n",
    "                with env.lock:\n",
    "                    mkpath(workdir)\n",
    "                #env.error(\"fid {} num {}\\n\".format(fid, fams[fid].get_member_ids()))\n",
    "                fam_af = af[(fid, s[1])]\n",
    "                if not fam_af:\n",
    "                    #env.log('All missing in this family {} on {}[{}], skipped ...'.format(fid, gene, gno), flush=True)\n",
    "                    with env.skipped_counter.get_lock():\n",
    "                        env.skipped_counter.value += 1\n",
    "                    removeEmptyDir(workdir)\n",
    "                    continue\n",
    "                ids = fams[fid].get_sorted_ids()\n",
    "                idxes = list(map(lambda x: fams[fid].get_member_idx(x), ids))\n",
    "                gs = list(map(lambda x: s[2 * x + 4 : 2 * x + 6], idxes))\n",
    "                gs_num = len(set(filter(lambda x: x != '0', chain(*gs))))\n",
    "                if gs_num >= 10:\n",
    "                    with env.skipped_counter.get_lock():\n",
    "                        env.skipped_counter.value += 1\n",
    "                    removeEmptyDir(workdir)\n",
    "                    continue\n",
    "                with env.lock:\n",
    "                    mkpath(workdir)\n",
    "                with open('{}/{}.PRE'.format(workdir, gno), 'w') as pre:\n",
    "                    pre.write(''.join(\"{} {} {} {}\\n\".format(fid, fams[fid].print_member(pid), s[2*fams[fid].get_member_idx(pid) + 4], s[2*fams[fid].get_member_idx(pid) + 5]) for pid in ids))\n",
    "                with open('{}/{}.LOC'.format(workdir, gno), 'w') as loc:\n",
    "                    loc.write(\"2 0 0 5\\n\")\n",
    "                    loc.write(\"0 0.0 0.0 0\\n\")\n",
    "                    loc.write(\"1 2\\n\")\n",
    "                    loc.write(\"1 2\\n\")\n",
    "                    loc.write(\" {} {}\\n\".format(1 - prev, prev))\n",
    "                    loc.write(\" 1\\n\")\n",
    "                    loc.write(\" {} {} {}\\n\".format(wild_pen, heter_pen, muta_pen))\n",
    "                    loc.write(\"3 {}\\n\".format(gs_num))\n",
    "                    loc.write(' ' + ' '.join(fam_af) + \"\\n\")\n",
    "                    loc.write(\"0 0\\n\")\n",
    "                    loc.write(\"0.0\\n\")\n",
    "                    loc.write(\"1 {} {}\\n\".format(theta_inc, theta_max))\n",
    "\n",
    "            removeEmptyDir('{}/{}'.format(out_base, gene))\n",
    "    tped_fh.close()\n",
    "    tfam_fh.close()\n",
    "    removeEmptyDir('{}'.format(out_base))\n",
    "\n",
    "#parse tfam file, store families into the Pedigree class\n",
    "def parse_tfam(fh):\n",
    "    fams = defaultdict(lambda: Pedigree())\n",
    "    idx = 0\n",
    "    for line in fh:\n",
    "        s = line.strip().split()\n",
    "        fams[s[0]].add_member(s[1:], idx)\n",
    "        idx += 1\n",
    "    return fams\n",
    "\n",
    "#This is to sort the members of a pedigree.\n",
    "#To make sure that parents come before offsprings.\n",
    "class Pedigree:\n",
    "    def __init__(self):\n",
    "        self.fid = None\n",
    "        self.data = {}\n",
    "        self.graph = defaultdict(lambda:[]) #, defaultdict(lambda:[])]\n",
    "        self.sorted = []\n",
    "\n",
    "    def add_member(self, info, idx): #list [pid, father, mother, sex, pheno]\n",
    "        if info[1] != '0' and info[2] != '0':\n",
    "            self.graph[info[1]].append(info[0])\n",
    "            self.graph[info[2]].append(info[0])\n",
    "        self.data[info[0]] = info + [idx]\n",
    "\n",
    "    def get_member_info(self, pid):\n",
    "        return self.data[pid][:-1]\n",
    "\n",
    "    def get_member_idx(self, pid):\n",
    "        return self.data[pid][-1]\n",
    "\n",
    "    def get_member_ids(self):\n",
    "        return self.data.keys()\n",
    "\n",
    "    def print_member(self, pid):\n",
    "        return ' '.join(self.get_member_info(pid))\n",
    "\n",
    "    def get_sorted_ids(self):\n",
    "        if self.sorted:\n",
    "            return self.sorted\n",
    "        else:\n",
    "            #This algorithm was first described by Kahn (1962)\n",
    "            S_no_parents = list(filter(lambda x: True if self.get_member_info(x)[1] == '0' else False, self.get_member_ids()))\n",
    "            graph = self.graph.copy()\n",
    "            while(S_no_parents):\n",
    "                n = S_no_parents.pop()\n",
    "                self.sorted.append(n)\n",
    "                if n not in graph:\n",
    "                    continue\n",
    "                offsprings = graph.pop(n)\n",
    "                for m in offsprings:\n",
    "                    father = self.get_member_info(m)[1]\n",
    "                    mother = self.get_member_info(m)[2]\n",
    "                    if father not in graph and mother not in graph:\n",
    "                        S_no_parents.append(m)\n",
    "            if graph:\n",
    "                raise Exception(\"There is a loop in the pedigree: {}\\n\".format(' '.join(graph.keys())))\n",
    "            else:\n",
    "                return self.sorted\n",
    "\n",
    "#runners\n",
    "def run_linkage(blueprint, theta_inc, theta_max, to_plot = True):\n",
    "    try:\n",
    "        remove_tree(os.path.join(env.outdir, 'heatmap'))\n",
    "    except OSError:\n",
    "        pass\n",
    "    with open(os.path.join(env.tmp_dir, 'LinkageRuntimeError.txt'), 'w') as runtime_err:\n",
    "        workdirs = glob.glob('{}/LINKAGE/{}.chr*'.format(env.tmp_dir, env.output))\n",
    "        parmap(lambda x: linkage_worker(blueprint, x, theta_inc, theta_max, runtime_err, to_plot) , workdirs, env.jobs)\n",
    "\n",
    "def linkage_worker(blueprint, workdir, theta_inc, theta_max, errfile, to_plot = True):\n",
    "    #env.log(\"Start running LINKAGE for {} ...\".format(workdir), flush=True)\n",
    "    #hash genes into genemap\n",
    "    genemap = {}\n",
    "    if blueprint:\n",
    "        with open(blueprint) as f:\n",
    "            for line in f.readlines():\n",
    "                chrID, start, end, gene = line.strip().split()[:4]\n",
    "                genemap[gene] = [chrID, int(start), int(end)]\n",
    "    else:\n",
    "        tped = os.path.join(env.tmp_cache, basename(workdir) + '.tped')\n",
    "        with open(tped) as f:\n",
    "            for line in f.readlines():\n",
    "                items = line.strip().split()[:4]\n",
    "                chrID = items[0]\n",
    "                gene = items[1]\n",
    "                pos = items[3]\n",
    "                genemap[gene] = [chrID, int(pos), int(pos)+1]\n",
    "    with env.lock:\n",
    "        mkpath('{}/heatmap'.format(env.outdir))\n",
    "    lods_fh = open('{}/heatmap/{}.lods'.format(env.outdir, basename(workdir)), 'w')\n",
    "    hlods_fh = open('{}/heatmap/{}.hlods'.format(env.outdir, basename(workdir)), 'w')\n",
    "    famlods_fh = open('{}/heatmap/{}.family-lods'.format(env.outdir, basename(workdir)), 'w')\n",
    "    genes = list(filter(lambda g: g in genemap, map(basename, glob.glob(workdir + '/*'))))\n",
    "    for gene in sorted(genes, key=lambda g: genemap[g]):\n",
    "        lods = {}\n",
    "        hlods = {}\n",
    "        fams = map(basename, filter(isdir, glob.glob('{}/{}/*'.format(workdir, gene))))\n",
    "        for fam in fams:\n",
    "            with cd('{}/{}/{}'.format(workdir, gene, fam)):\n",
    "                units = map(lambda x: re.sub(r'^(\\d+?)\\.PRE$', r'\\1', x) ,glob.glob('*.PRE'))\n",
    "                for unit in units:\n",
    "                    copy_file('{}.LOC'.format(unit), 'datafile.dat')\n",
    "                    copy_file('{}.PRE'.format(unit), 'pedfile.pre')\n",
    "\n",
    "                    step1 = runCommand(['makeped', 'pedfile.pre', 'pedfile.ped', 'n'],\n",
    "                                       show_stderr = False, return_zero = False)\n",
    "                    if step1[1]:\n",
    "                        if env.debug:\n",
    "                            with env.lock:\n",
    "                                print(\"makeped error:\", step1[1], file = sys.stderr)\n",
    "                        with env.makeped_counter.get_lock():\n",
    "                            env.makeped_counter.value += 1\n",
    "                        with env.lock:\n",
    "                            errfile.write(step1[1])\n",
    "                        continue\n",
    "                    step2 = runCommand(['pedcheck', '-p', 'pedfile.ped', '-d', 'datafile.dat', '-c'],\n",
    "                                       show_stderr = False, return_zero = False)\n",
    "                    if step2[1]:\n",
    "                        lines = [x for x in step2[1].split('\\n')\n",
    "                                 if not x.strip().startswith('*') and x.strip()]\n",
    "                        if len(lines) > 0:\n",
    "                            env.log('{} lines'.format(len(lines)))\n",
    "                            with env.lock:\n",
    "                                errfile.write(step2[1])\n",
    "                            if env.debug:\n",
    "                                with env.lock:\n",
    "                                    print(\"pedcheck error:\", '\\n'.join(lines), file = sys.stderr)\n",
    "                    try:\n",
    "                        copy_file('zeroout.dat', 'pedfile.dat')\n",
    "                    except:\n",
    "                        with env.pedcheck_counter.get_lock():\n",
    "                            env.pedcheck_counter.value += 1\n",
    "                        continue\n",
    "                    step3 = runCommand('unknown', show_stderr = False, return_zero = False)\n",
    "                    if step3[1]:\n",
    "                        if env.debug:\n",
    "                            with env.lock:\n",
    "                                print(\"unkn error:\", step3[1], file = sys.stderr)\n",
    "                        with env.unknown_counter.get_lock():\n",
    "                            env.unknown_counter.value += 1\n",
    "                        with env.lock:\n",
    "                            errfile.write(step3[1])\n",
    "                        continue\n",
    "                    step4 = runCommand('mlink', show_stderr = False, return_zero = False)\n",
    "                    if step4[1]:\n",
    "                        if env.debug:\n",
    "                            with env.lock:\n",
    "                                print(\"mlink error:\", step4[1], file = sys.stderr)\n",
    "                        with env.lock:\n",
    "                            errfile.write(step4[1])\n",
    "                    try:\n",
    "                        copy_file('outfile.dat', '{}.out'.format(unit))\n",
    "                    except:\n",
    "                        with env.mlink_counter.get_lock():\n",
    "                            env.mlink_counter.value += 1\n",
    "                        continue\n",
    "                    #clean linkage tmp files\n",
    "                    for f in set(glob.glob('*.dat') + glob.glob('ped*') + ['names.tmp']):\n",
    "                        os.remove(f)\n",
    "                    #collect lod scores of different thelta for the fam\n",
    "                    with open('{}.out'.format(unit)) as out:\n",
    "                        raw = out.read()\n",
    "                        for i in re.finditer(r'^THETAS\\s+(0\\.\\d+)(?:\\n.+?){7}LOD SCORE =\\s+(-?\\d+\\.\\d+)', raw, re.MULTILINE):\n",
    "                            theta, lod = map(float, i.group(1,2))\n",
    "                            #if float(lod) < 1e-6:\n",
    "                            #    lod = 0\n",
    "                            if theta not in lods:\n",
    "                                lods[theta] = {fam: lod}\n",
    "                            elif fam not in lods[theta] or lod > lods[theta][fam]:\n",
    "                                lods[theta][fam] = lod\n",
    "        for theta in sorted(lods.keys()):\n",
    "            lods_fh.write('{} {} {} {}\\n'.format(gene, ' '.join(map(str, genemap[gene])), theta, sum(lods[theta].values())))\n",
    "            for fam in lods[theta]:\n",
    "                famlods_fh.write('{} {} {} {} {}\\n'.format(fam, gene, ' '.join(map(str, genemap[gene])), theta, lods[theta][fam]))\n",
    "            res = minimize_scalar(hlod_fun(lods[theta].values(), -1), bounds=(0,1), method='bounded', options={'xatol':1e-8})\n",
    "            a = res.x\n",
    "            hlods_fh.write('{} {} {} {} {}\\n'.format(gene, ' '.join(map(str, genemap[gene])), a, theta, hlod_fun(lods[theta].values())(a)))\n",
    "        with env.run_counter.get_lock():\n",
    "            env.run_counter.value += 1\n",
    "        if env.run_counter.value % (env.batch * env.jobs) == 0:\n",
    "            env.log('Linkage analysis for {:,d} units completed {{{:.2%}}} ...'.format(env.run_counter.value, float(env.run_counter.value)/env.success_counter.value), flush=True)\n",
    "    lods_fh.close()\n",
    "    hlods_fh.close()\n",
    "    famlods_fh.close()\n",
    "    if to_plot:\n",
    "        heatmap('{}/heatmap/{}.lods'.format(env.outdir, basename(workdir)), theta_inc, theta_max)\n",
    "        heatmap('{}/heatmap/{}.hlods'.format(env.outdir, basename(workdir)), theta_inc, theta_max)\n",
    "    #env.log(\"Finished running LINKAGE for {}.\".format(workdir), flush=True)\n",
    "\n",
    "\n",
    "def hinton(filename, max_weight=None, ax=None):\n",
    "    if ax is None:\n",
    "        ax = plt.gca()\n",
    "    matrix = np.random.rand(20, 20) - 0.5\n",
    "    if not max_weight:\n",
    "        max_weight = 2**np.ceil(np.log(np.abs(matrix).max())/np.log(2))\n",
    "    ax.patch.set_facecolor('gray')\n",
    "    ax.set_aspect('equal', 'box')\n",
    "    ax.xaxis.set_major_locator(plt.NullLocator())\n",
    "    ax.yaxis.set_major_locator(plt.NullLocator())\n",
    "    chrID = re.search(r'\\.chr([0-9XY]+)\\.', filename).group(1)\n",
    "    ax.set_title('Chromosome {}'.format(chrID))\n",
    "    for (x,y),w in np.ndenumerate(matrix):\n",
    "        color = 'white' if w > 0 else 'black'\n",
    "        size = np.sqrt(np.abs(w))\n",
    "        rect = plt.Rectangle([x - size / 2, y - size / 2], size, size,\n",
    "                             facecolor=color, edgecolor=color)\n",
    "        ax.add_patch(rect)\n",
    "    ax.autoscale_view()\n",
    "    ax.invert_yaxis()\n",
    "    plt.savefig(filename)\n",
    "\n",
    "\n",
    "def heatmap(file, theta_inc, theta_max):\n",
    "    #env.log(\"Start ploting heatmap for {} ...\".format(file), flush=True)\n",
    "    if os.path.getsize(file) == 0:\n",
    "        hinton('{}.png'.format(file))\n",
    "        return\n",
    "    lods = []\n",
    "    with open(file, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            theta,lod = line.split()[-2:]\n",
    "            if float(theta) >= theta_max:\n",
    "                continue\n",
    "            lods.append(lod)\n",
    "        if max(lods) == min(lods):\n",
    "            #env.log('Max equals Min for [{}], No real heatmap will be generated.'.format(file))\n",
    "            hinton('{}.png'.format(file))\n",
    "            return\n",
    "        Num=int(round(theta_max/theta_inc))\n",
    "        lods = np.array(map(float,lods)).reshape((-1,Num))\n",
    "        chrID = re.search(r'\\.chr([0-9XY]+)\\.', file).group(1)\n",
    "        fig, ax = plt.subplots(1)\n",
    "        ax.set_title('Chromosome {}'.format(chrID))\n",
    "        plt.pcolormesh(fig,ax,lods.transpose(),\n",
    "                       xticklabels=[''] * len(lods),\n",
    "                       yticklabels=np.round(np.array(range(Num)) * theta_inc,2).tolist(),\n",
    "                       cmap=brewer2mpl.get_map('Blues', 'Sequential', 9).mpl_colormap)\n",
    "        fig.savefig('{}.png'.format(file))\n",
    "        #fig.close()\n",
    "    #env.log(\"Finished ploting heatmap for {}.\".format(file), flush=True)\n",
    "\n",
    "def hlod_fun(Li, sign=1):\n",
    "    def _fun(alpha):\n",
    "        return sign * sum(np.log10(alpha*np.power(10, Li) + 1 - alpha))\n",
    "    return _fun\n",
    "\n",
    "def html(theta_inc, theta_max, limit):\n",
    "    if limit <= 0:\n",
    "        return\n",
    "    head = \"\"\"<html>\n",
    "    <head>\n",
    "    <title>Results for {}</title>\"\"\".format(env.output) + \"\"\"\n",
    "    <style>\n",
    "    table {\n",
    "\tfont-family: verdana,arial,sans-serif;\n",
    "\tfont-size:12px;\n",
    "\tcolor:#333333;\n",
    "\tborder-width: 1px;\n",
    "\tborder-color: #CCCCCC;\n",
    "\tborder-collapse: collapse;\n",
    "    }\n",
    "    table th {\n",
    "\tborder-width: 1px;\n",
    "\tpadding: 6px;\n",
    "\tcolor: #006295;\n",
    "\tborder-style: solid;\n",
    "\tborder-color: #CCCCCC;\n",
    "\tbackground-color: #F8F8F8;\n",
    "    }\n",
    "    table td {\n",
    "\tborder-width: 1px;\n",
    "\tpadding: 6px;\n",
    "\tborder-style: solid;\n",
    "\tborder-color: #CCCCCC;\n",
    "\tbackground-color: #ffffff;\n",
    "    }\n",
    "    a,\n",
    "    a:link,\n",
    "    a:visited,\n",
    "    a:active\n",
    "    {\n",
    "\tcolor: #3366CC;\n",
    "\tborder-bottom: 1px dotted #3366CC;\n",
    "\ttext-decoration:none;\n",
    "    }\n",
    "    a:hover\n",
    "    {\n",
    "\tborder-bottom: none;\n",
    "\tcolor: #ffffff;\n",
    "    background-color: #006295;\n",
    "    }\n",
    "    body {font-family:Lucida Sans Unicode,arial,sans-serif;}\n",
    "    </style>\n",
    "    <script type=\"text/javascript\">\n",
    "    function toggle(obj) {\n",
    "    var elstyle = document.getElementById(obj).style;\n",
    "    var text    = document.getElementById(obj + \"tog\");\n",
    "    if (elstyle.display == \\'none\\') {\n",
    "    elstyle.display = \\'block\\';\n",
    "    text.innerHTML = \"hide\";\n",
    "    } else {\n",
    "    elstyle.display = \\'none\\';\n",
    "    text.innerHTML = \"show\";\n",
    "    }\n",
    "    }</script>\n",
    "    </head>\"\"\"\n",
    "    body = \"\"\"<body>\n",
    "    <p><a href=\"#Lods_Table\" onclick=\"toggle(\\'lods_tbl\\')\">Ranked LOD scores</a>\n",
    "    <div id=\"lods_tbl\" class=\"divinfo\", style=\"border:0px;width:auto;height:auto;overflow-y:hidden;overflow-x:scroll;\">{}</div></p>\n",
    "    <p><a href=\"#Hlods_Table\" onclick=\"toggle(\\'hlods_tbl\\')\">Ranked HLOD scores</a>\n",
    "    <div id=\"hlods_tbl\" class=\"divinfo\", style=\"border:0px;width:auto;height:auto;overflow-y:hidden;overflow-x:scroll;\">{}</div></p>\n",
    "    <p><a href=\"#Lods_Heatmap\" onclick=\"toggle(\\'lods_heatmap\\')\">LOD scores heatmap</a>\n",
    "    <div id=\"lods_heatmap\">{}</div></p>\n",
    "    <p><a href=\"#Hlods_Heatmap\" onclick=\"toggle(\\'hlods_heatmap\\')\">HLOD scores heatmap</a>\n",
    "    <div id=\"hlods_heatmap\">{}</div></p>\n",
    "    </body>\n",
    "    </html>\"\"\"\n",
    "    env.log('Generating Report in HTML format ...', flush = True)\n",
    "    with open('{}/{}_Report.html'.format(env.outdir, env.output), 'w') as f:\n",
    "        #t = Template(index)\n",
    "        #c = Context({ \"lods\": lods_tbl })\n",
    "        f.write(head + body.format(html_table('Lod', theta_inc, theta_max, limit), html_table('Hlod', theta_inc, theta_max, limit), html_img('lod'), html_img('hlod')))\n",
    "    env.log('Report for [{}] is generated in HTML format\\n'.format(env.output), flush = True)\n",
    "\n",
    "def html_img(ltype):\n",
    "    chrs = ['{}'.format(i+1) for i in range(22)] + ['X', 'Y']\n",
    "    imgs = ['heatmap/{0}.chr{1}.{2}s.png'.format(env.output, chrID, ltype) for chrID in chrs]\n",
    "    exist_imgs = [img for img in imgs if isfile('{}/{}'.format(env.outdir,img))]\n",
    "    return ''.join('<img src={}></img>'.format(img) for img in exist_imgs)\n",
    "\n",
    "def html_table(type, theta_inc, theta_max, limit):\n",
    "    colNum = int(round(theta_max/theta_inc))\n",
    "    theta_orders = range(colNum)\n",
    "    thetas = map(lambda x: theta_inc * x, range(colNum))\n",
    "    #table\n",
    "    table = r'<table style=\"width:300px;font-size:12px\">{}</table>'\n",
    "    #table header\n",
    "    lods_header = r'<tr>{}</tr>'.format(''.join(r'<th colspan=\"2\">&#952;={}</td>'.format(x) for x in thetas))\n",
    "    lods_header += r'<tr>{}</tr>'.format(r'<th rowspan=\"2\">{}</th><th>Marker name</th>'.format(type) * colNum)\n",
    "    lods_header += r'<tr>{}</tr>'.format(r'<th>chr:start-end</th>' * colNum)\n",
    "    #initialize lods dict\n",
    "    lods = {}\n",
    "    for theta_order in theta_orders:\n",
    "        lods[theta_order] = {}\n",
    "        #print '{}\\n'.format(theta_order)\n",
    "    #read lods\n",
    "    lods_files = glob.glob('{}/heatmap/{}.*.{}s'.format(env.outdir, env.output, type.lower()))\n",
    "    for file in lods_files:\n",
    "        with open(file, 'r') as f:\n",
    "            for line in f:\n",
    "                if type == 'Lod':\n",
    "                    gene, chrId, start, end, theta, lod = line.strip().split()\n",
    "                    if int(round(float(theta)/theta_inc)) >= colNum:\n",
    "                        continue\n",
    "                    lods[int(round(float(theta)/theta_inc))][gene] = ['<b>{}</b>'.format(round(float(lod),3)), gene, chrId, start, end]\n",
    "                elif type == 'Hlod':\n",
    "                    gene, chrId, start, end, a, theta, lod = line.strip().split()\n",
    "                    if int(round(float(theta)/theta_inc)) >= colNum:\n",
    "                        continue\n",
    "                    lods[int(round(float(theta)/theta_inc))][gene] = ['<b>{}</b><br>&#945;={}'.format(round(float(lod),3), round(float(a),3)), gene, chrId, start, end]\n",
    "                else:\n",
    "                    env.error('Wrong type of LOD')\n",
    "\n",
    "    #collect result\n",
    "    res = np.empty((len(theta_orders), limit), dtype=list)\n",
    "    for theta_order in theta_orders:\n",
    "        i=0\n",
    "        for gene in sorted(lods[theta_order].keys(), key=lambda x: lods[theta_order][x][0], reverse=True):\n",
    "            if i >= limit:\n",
    "                break\n",
    "            res[theta_order][i] = lods[theta_order][gene]\n",
    "            i += 1\n",
    "    #print res\n",
    "    #write lods table\n",
    "    lods_res = ''\n",
    "    for i in range(min(limit, len(lods[0].keys()))):\n",
    "        lods_res += r'<tr>{}</tr>'.format(''.join(r'<td rowspan=\"2\">{}</td><td>{}</td>'.format(res[theta_order][i][0],res[theta_order][i][1]) for theta_order in theta_orders))\n",
    "        lods_res += r'<tr>{}</tr>'.format(''.join(r'<td>{}:{}-{}</td>'.format(res[theta_order][i][2],res[theta_order][i][3],res[theta_order][i][4]) for theta_order in theta_orders))\n",
    "    lods_tbl = table.format(lods_header + lods_res)\n",
    "    return lods_tbl\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
