{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp Main"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main module\n",
    "\n",
    "> the main function of SEQLinkage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "from argparse import ArgumentParser, ArgumentTypeError, RawDescriptionHelpFormatter, SUPPRESS\n",
    "import os, glob, platform\n",
    "from multiprocessing import cpu_count, Queue\n",
    "from SEQLinkage.Utils import *\n",
    "from SEQLinkage.Runner import *\n",
    "from SEQLinkage.Core import *\n",
    "from multiprocessing import Process, Queue\n",
    "from collections import OrderedDict\n",
    "import itertools\n",
    "from copy import deepcopy\n",
    "import sys, faulthandler, platform\n",
    "import numpy as np\n",
    "import os\n",
    "if sys.version_info.major == 2:\n",
    "    from cstatgen import cstatgen_py2 as cstatgen\n",
    "    from cstatgen.egglib import Align\n",
    "else:\n",
    "    from cstatgen import cstatgen_py3 as cstatgen\n",
    "    import egglib\n",
    "    from egglib import Align\n",
    "\n",
    "class Args:\n",
    "    def __init__(self):\n",
    "        self.parser = ArgumentParser(description = '''\\t{}, linkage analysis using sequence data\\n\\t'''.format(\"SEQLinkage V\"+VERSION))\n",
    "        self.getEncoderArguments(self.parser)\n",
    "        self.getLinkageArguments(self.parser)\n",
    "\n",
    "\n",
    "    def get(self):\n",
    "        return self.parser.parse_args()\n",
    "\n",
    "    def getEncoderArguments(self, parser):\n",
    "        vargs = parser.add_argument_group('Collapsed haplotype pattern method arguments') \n",
    "        vargs.add_argument('--chp', action='store_true', help='''Generate CHP markers.''')\n",
    "        vargs.add_argument('--fam', metavar='FILE', required=True, dest = \"tfam\", help='''Input pedigree and phenotype information in FAM format.''')\n",
    "        vargs.add_argument('--vcf', metavar='FILE', required=True, help='''Input VCF file, bgzipped.''')\n",
    "        vargs.add_argument('--anno', metavar='FILE', required=False, help='''Input annotation file from annovar.''')\n",
    "        vargs.add_argument('--pop', metavar='FILE', required=False, help='''Input two columns file, first column is family ID, second column population information.''')\n",
    "        vargs.add_argument('--included-vars', metavar='FILE', dest='included_vars', help='''Variants to be included for linkage analysis, if None, the analysis won't filter any variants. But you can still set AF cutoff by -c argment.''')\n",
    "        vargs.add_argument('-b', '--blueprint', metavar = 'FILE',\n",
    "                           help='''Blueprint file that defines regional marker (format: \"chr startpos endpos name avg.distance male.distance female.distance\").''')\n",
    "        vargs.add_argument('-c', '--maf-cutoff', metavar='P', default=1.0, type=float, dest = \"maf_cutoff\",\n",
    "                           help='''MAF cutoff to define variants to be excluded from analyses. this is useful, if you need to analyse multiple population together.''')\n",
    "        vargs.add_argument('-o', '--output', metavar='Name', default='data/linkage_analysis/', help='''Output name prefix.''')\n",
    "        \n",
    "        vargs.add_argument('--build', metavar='STRING', default='hg38', choices = [\"hg19\", \"hg38\"], help='''Reference genome version for VCF file.''')\n",
    "        vargs.add_argument('--freq', metavar='INFO', default = 'AF',help='''Info field name for allele frequency in VCF file.''')\n",
    "        vargs.add_argument('--chrom-prefix', metavar='STRING', dest = 'chr_prefix',\n",
    "                           help='''Prefix to chromosome name in VCF file if applicable, e.g. \"chr\".''')\n",
    "\n",
    "    def getLinkageArguments(self, parser):\n",
    "        vargs = parser.add_argument_group('LINKAGE options')\n",
    "        vargs.add_argument('--run-linkage', action='store_true', dest = \"run_linkage\",\n",
    "                           help='''Perform Linkage analysis.''')\n",
    "        vargs.add_argument('-K', '--prevalence', metavar='FLOAT', type=float,default = 0.001,\n",
    "                           help='Disease prevalence. Default to 0.001.')\n",
    "        vargs.add_argument('--moi', metavar='STRING', dest = \"inherit_mode\",default = 'AD',\n",
    "                           # choices=['AD', 'AR', 'Xlinked', 'Y'],\n",
    "                           choices=['AD', 'AR'],\n",
    "                           help='Mode of inheritance, AD/AR: autosomal dominant/recessive. Default to AD.')\n",
    "        vargs.add_argument('-W', '--wt-pen', metavar='FLOAT', type=float, dest = \"wild_pen\",default=0.01,\n",
    "                           help='Penetrance for wild type. Default to 0.01.')\n",
    "        vargs.add_argument('-M', '--mut-pen', metavar='FLOAT', type=float, dest = \"muta_pen\",default=0.9,\n",
    "                           help='Penetrance for mutation. Default to 0.9.')\n",
    "        vargs.add_argument('--theta-max', metavar='FLOAT', type=float, dest = \"theta_max\", default = 0.5,\n",
    "                           help='Theta upper bound. Default to 0.5.')\n",
    "        vargs.add_argument('--theta-inc', metavar='FLOAT', type=float, dest = \"theta_inc\", default = 0.05,\n",
    "                           help='Theta increment. Default to 0.05.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def checkParams(args):\n",
    "    '''set default arguments or make warnings'''\n",
    "    env.setoutput(args.output)\n",
    "    args.vcf = os.path.abspath(os.path.expanduser(args.vcf))\n",
    "    args.tfam = os.path.abspath(os.path.expanduser(args.tfam))\n",
    "    for item in [args.vcf, args.tfam]:\n",
    "        if not os.path.exists(item):\n",
    "            env.error(\"Cannot find file [{}]!\".format(item), exit = True)\n",
    "    if len([x for x in set(getColumn(args.tfam, 6)) if x.lower() not in env.ped_missing]) > 2:\n",
    "        env.trait = 'quantitative'\n",
    "    env.log('{} trait detected in [{}]'.format(env.trait.capitalize(), args.tfam))\n",
    "    if not args.blueprint:\n",
    "        if not args.anno:\n",
    "            args.blueprint = os.path.join(env.resource_dir, 'genemap.{}.txt'.format(args.build))\n",
    "        else:\n",
    "            env.log('Generate regions by annotation')\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def main():\n",
    "    '''the main encoder function'''\n",
    "    args = Args().get()\n",
    "    checkParams(args)   \n",
    "    env.log(args)\n",
    "    # load data\n",
    "    data = RData(args.vcf, args.tfam,args.anno,args.pop,allele_freq_info=args.freq,included_variant_file=args.included_vars)\n",
    "    samples_vcf = data.samples_vcf\n",
    "\n",
    "    if len(samples_vcf) == 0:\n",
    "        env.error(\"Fail to extract samples from [{}]\".format(args.vcf), exit = True)\n",
    "    env.log('{:,d} samples found in [{}]'.format(len(samples_vcf), args.vcf))\n",
    "    samples_not_vcf = data.samples_not_vcf\n",
    "\n",
    "    if len(data.families) == 0:\n",
    "        env.error('No valid family to process. ' \\\n",
    "                  'Families have to be at least trio with at least one member in VCF file.', exit = True)\n",
    "    if len(data.samples) == 0:\n",
    "        env.error('No valid sample to process. ' \\\n",
    "                  'Samples have to be in families, and present in both TFAM and VCF files.', exit = True)\n",
    "\n",
    "    if args.blueprint:\n",
    "        # load blueprint\n",
    "        try:\n",
    "            env.log('Loading marker map from [{}] ...'.format(args.blueprint))\n",
    "            with open(args.blueprint, 'r') as f:\n",
    "                regions = [x.strip().split() for x in f.readlines()]\n",
    "        except IOError:\n",
    "            env.error(\"Cannot load regional marker blueprint [{}]. \".format(args.blueprint), exit = True)\n",
    "    else:\n",
    "        env.log('separate chromosome to regions')\n",
    "        regions=data.get_regions(step=1000)  #whole-genome linkage analysis\n",
    "    env.log('{:,d} families with a total of {:,d} samples will be scanned for {:,d} pre-defined units'.\\\n",
    "            format(len(data.families), len(data.samples), len(regions)))\n",
    "  \n",
    "    run_each_region(regions,data,RegionExtractor(args.vcf, build = args.build, chr_prefix = args.chr_prefix),MarkerMaker(maf_cutoff = args.maf_cutoff),LinkageWriter(len(samples_not_vcf)),\n",
    "                    runlinkage=args.run_linkage,cutoff=args.maf_cutoff,chp=args.chp,rho=np.arange(0,args.theta_max,args.theta_inc),\n",
    "                    model = args.inherit_mode,penetrances = [args.wild_pen,args.muta_pen,args.muta_pen],dfreq=args.prevalence)\n",
    "    env.log('============= Finish analysis ==============')\n",
    "      \n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
