{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp Runner"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Runner module\n",
    "\n",
    "> API details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "from __future__ import print_function\n",
    "import os\n",
    "from SEQLinkage.Utils import *\n",
    "from collections import deque, defaultdict\n",
    "from os.path import splitext, basename, isdir, isfile\n",
    "from itertools import chain\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import brewer2mpl\n",
    "from scipy.optimize import minimize_scalar\n",
    "import re, glob\n",
    "from shutil import rmtree as remove_tree\n",
    "from distutils.file_util import copy_file\n",
    "#formatters\n",
    "#the handler, called from main, can call specific formatter.\n",
    "def format(tpeds, tfam, prev = None, wild_pen = None, muta_pen = None, out_format = 'MERLIN', inherit_mode = None, theta_max = None, theta_inc = None):\n",
    "    print(env.jobs) #testing line\n",
    "    if out_format == 'plink':\n",
    "        parmap(lambda x: format_plink(x, tfam), tpeds, env.jobs)\n",
    "    elif out_format == 'mega2':\n",
    "        mkpath(os.path.join(env.outdir, 'MEGA2'))\n",
    "        parmap(lambda x: format_mega2(x, tfam), tpeds, env.jobs)\n",
    "    elif out_format == 'merlin':\n",
    "        mkpath(os.path.join(env.outdir, 'MERLIN'))\n",
    "        parmap(lambda x: format_merlin(x, tfam), tpeds, env.jobs)\n",
    "    elif out_format == 'linkage':\n",
    "        parmap(lambda x: format_linkage(x, tfam, prev, wild_pen, muta_pen, inherit_mode, theta_max, theta_inc), tpeds, env.jobs)\n",
    "\n",
    "#plink format, ped and map\n",
    "def format_plink(tped, tfam):\n",
    "    out_base = '{}/PLINK/{}'.format(env.outdir, splitext(basename(tped))[0])\n",
    "    with open(tped) as tped_fh, open(tfam) as tfam_fh:\n",
    "        geno = []\n",
    "        with open(out_base + '.map', 'w') as m:\n",
    "            for line in tped_fh:\n",
    "                s = line.strip().split()\n",
    "                m.write(env.delimiter.join(s[:3]) + '\\n')\n",
    "                geno.append(deque(s[4:]))\n",
    "        with open(out_base + '.ped', 'w') as p:\n",
    "            for line in tfam_fh:\n",
    "                p.write(line.strip())\n",
    "                list(map(lambda x: p.write(' {} {}'.format(x.popleft(), x.popleft)), geno))\n",
    "                p.write(\"\\n\")\n",
    "\n",
    "#mega2 format, datain.01, pedin.01, map.01\n",
    "def format_mega2(tped, tfam):\n",
    "    trait = 'A' if env.trait == 'binary' else 'T'\n",
    "    pedheader = ['Pedigree', 'ID', 'Father', 'Mother', 'Sex', 'Trait.{}'.format(trait)]\n",
    "    out_base = os.path.join(env.outdir, 'MEGA2')\n",
    "    suffix =  re.search(r'chr([0-9XY]+)', basename(tped)).groups()[0]\n",
    "    with open(tped) as tph, open(tfam) as tfh:\n",
    "        geno = []\n",
    "        name = []\n",
    "        with open('{}/map.{}'.format(out_base, suffix), 'w') as m,\\\n",
    "             open('{}/datain.{}'.format(out_base, suffix), 'w') as d:\n",
    "            d.write('Type\\tName\\n{}\\tTrait\\n'.format(trait))\n",
    "            m.write('Chromosome\\tMap.k.a\\tName\\tMap.k.m\\tMap.k.f\\thg19.p\\n')\n",
    "            for line in tph:\n",
    "                s = line.strip().split()\n",
    "                gene, gno = re.search(r'^(\\S+?)(?:\\[(\\d+)\\])?$', s[1]).groups()\n",
    "                if not gno:\n",
    "                    gno = '0'\n",
    "                    with env.format_counter.get_lock():\n",
    "                        env.format_counter.value += 1\n",
    "                elif gno == '1':\n",
    "                    with env.format_counter.get_lock():\n",
    "                        env.format_counter.value += 1\n",
    "                if env.format_counter.value % (env.batch * env.jobs) == 0:\n",
    "                    env.log('{:,d} units processed {{{:.2%}}} ...'.format(env.format_counter.value, float(env.format_counter.value)/env.success_counter.value), flush=True)\n",
    "                d.write('M\\t{}\\n'.format(s[1]))\n",
    "                dis = s[2].split(';')\n",
    "                dis.insert(1,s[1])\n",
    "                m.write('{}\\t{}\\t{}\\n'.format(s[0],'\\t'.join(dis),s[3]))\n",
    "                name.append('{0}.M.1\\t{0}.M.2'.format(s[1]))\n",
    "                geno.append(deque(map(lambda x: re.sub(r'^0$', 'NA', x), s[4:])))\n",
    "        with open('{}/pedin.{}'.format(out_base, suffix), 'w') as p:\n",
    "            p.write('{}\\n'.format('\\t'.join(pedheader + name)))\n",
    "            for line in tfh:\n",
    "                p.write(line.strip())\n",
    "                s = line.strip().split()\n",
    "                list(map(lambda x: p.write('\\t{}\\t{}'.format(x.popleft(), x.popleft())), geno))\n",
    "                p.write(\"\\n\")\n",
    "\n",
    "#merlin format\n",
    "def format_merlin(tped, tfam):\n",
    "    trait = 'A' if env.trait == 'binary' else 'T'\n",
    "    #pedheader = ['Pedigree', 'ID', 'Father', 'Mother', 'Sex', 'Trait.{}'.format(trait)]\n",
    "    out_base = os.path.join(env.outdir, 'MERLIN', splitext(basename(tped))[0])\n",
    "    #suffix =  re.search(r'chr([0-9XY]+)', basename(tped)).groups()[0]\n",
    "    with open(tped) as tph, open(tfam) as tfh:\n",
    "        geno = []\n",
    "        #name = []\n",
    "        with open('{}.map'.format(out_base), 'w') as m,\\\n",
    "             open('{}.dat'.format(out_base), 'w') as d:\n",
    "            d.write('{}\\tTrait\\n'.format(trait))\n",
    "            m.write('CHROMOSOME\\tMARKER\\t\\tPOSITION\\tFEMALE_POSITION\\tMALE_POSITION\\n')\n",
    "            for line in tph:\n",
    "                s = line.strip().split()\n",
    "                gene, gno = re.search(r'^(\\S+?)(?:\\[(\\d+)\\])?$', s[1]).groups()\n",
    "                if not gno:\n",
    "                    gno = '0'\n",
    "                    with env.format_counter.get_lock():\n",
    "                        env.format_counter.value += 1\n",
    "                elif gno == '1':\n",
    "                    with env.format_counter.get_lock():\n",
    "                        env.format_counter.value += 1\n",
    "                if env.format_counter.value % (env.batch * env.jobs) == 0:\n",
    "                    env.log('{:,d} units processed {{{:.2%}}} ...'.format(env.format_counter.value, float(env.format_counter.value)/env.success_counter.value), flush=True)\n",
    "                d.write('M\\t{}\\n'.format(s[1]))\n",
    "                dis = s[2].split(';')\n",
    "                dis[1], dis[2] = dis[2], dis[1]\n",
    "                #dis.insert(1,s[1])\n",
    "                m.write('{}\\t{}\\t{}\\n'.format(s[0],s[1],'\\t'.join(dis)))\n",
    "                #name.append('{0}.M.1\\t{0}.M.2'.format(s[1]))\n",
    "                geno.append(deque(s[4:]))\n",
    "        with open('{}.ped'.format(out_base), 'w') as p:\n",
    "            #p.write('{}\\n'.format('\\t'.join(pedheader + name)))\n",
    "            for line in tfh:\n",
    "                p.write(line.strip())\n",
    "                s = line.strip().split()\n",
    "                list(map(lambda x: p.write('\\t{}\\t{}'.format(x.popleft(), x.popleft())), geno))\n",
    "                p.write(\"\\n\")\n",
    "\n",
    "\n",
    "#linkage format, .pre and .loc\n",
    "#per locus, per family based\n",
    "#because the haplotype patterns are different from family to family.\n",
    "#You can analyze them all together\n",
    "def format_linkage(tped, tfam, prev, wild_pen, muta_pen, inherit_mode, theta_max, theta_inc):\n",
    "    out_base = '{}/LINKAGE/{}'.format(env.tmp_dir, splitext(basename(tped))[0])\n",
    "    with open(tped) as tped_fh, open(tfam) as tfam_fh:\n",
    "        fams = parse_tfam(tfam_fh)\n",
    "        #parse per family per locus AF file\n",
    "        af = defaultdict(lambda: [])\n",
    "        #try to open the file for allele frequencies, otherwise use the default value\n",
    "        try:\n",
    "            with open(os.path.join(env.tmp_cache, basename(out_base) + '.freq')) as af_fh:\n",
    "                for line in af_fh:\n",
    "                    s = line.strip().split()\n",
    "                    freq = list(map(lambda x: max(1e-3, float(x)), s[2:]))\n",
    "                    relativefreq = np.array(freq)/sum(freq)\n",
    "                    af[(s[0],s[1])] = list(map(str, relativefreq))\n",
    "        except IOError:\n",
    "            env.error('freq info not properly read for [{}]'.format(basename(out_base)))\n",
    "        #parse tped\n",
    "        heter_pen = wild_pen\n",
    "        if inherit_mode == 'AD':\n",
    "            heter_pen = muta_pen\n",
    "        for line in tped_fh:\n",
    "            s = line.strip().split()\n",
    "            gene, gno = re.search(r'^(\\S+?)(?:\\[(\\d+)\\])?$', s[1]).groups()\n",
    "            if not gno:\n",
    "                gno = '0'\n",
    "                with env.format_counter.get_lock():\n",
    "                    env.format_counter.value += 1\n",
    "            elif gno == '1':\n",
    "                with env.format_counter.get_lock():\n",
    "                    env.format_counter.value += 1\n",
    "            if env.format_counter.value % (env.batch * env.jobs) == 0:\n",
    "                env.log('{:,d} units processed {{{:.2%}}} ...'.format(env.format_counter.value, float(env.format_counter.value)/env.success_counter.value), flush=True)\n",
    "            for fid in fams:\n",
    "                workdir = '{}/{}/{}'.format(out_base, gene, fid)\n",
    "                with env.lock:\n",
    "                    mkpath(workdir)\n",
    "                #env.error(\"fid {} num {}\\n\".format(fid, fams[fid].get_member_ids()))\n",
    "                fam_af = af[(fid, s[1])]\n",
    "                if not fam_af:\n",
    "                    #env.log('All missing in this family {} on {}[{}], skipped ...'.format(fid, gene, gno), flush=True)\n",
    "                    with env.skipped_counter.get_lock():\n",
    "                        env.skipped_counter.value += 1\n",
    "                    removeEmptyDir(workdir)\n",
    "                    continue\n",
    "                ids = fams[fid].get_sorted_ids()\n",
    "                idxes = list(map(lambda x: fams[fid].get_member_idx(x), ids))\n",
    "                gs = list(map(lambda x: s[2 * x + 4 : 2 * x + 6], idxes))\n",
    "                gs_num = len(set(filter(lambda x: x != '0', chain(*gs))))\n",
    "                if gs_num >= 10:\n",
    "                    with env.skipped_counter.get_lock():\n",
    "                        env.skipped_counter.value += 1\n",
    "                    removeEmptyDir(workdir)\n",
    "                    continue\n",
    "                with env.lock:\n",
    "                    mkpath(workdir)\n",
    "                with open('{}/{}.PRE'.format(workdir, gno), 'w') as pre:\n",
    "                    pre.write(''.join(\"{} {} {} {}\\n\".format(fid, fams[fid].print_member(pid), s[2*fams[fid].get_member_idx(pid) + 4], s[2*fams[fid].get_member_idx(pid) + 5]) for pid in ids))\n",
    "                with open('{}/{}.LOC'.format(workdir, gno), 'w') as loc:\n",
    "                    loc.write(\"2 0 0 5\\n\")\n",
    "                    loc.write(\"0 0.0 0.0 0\\n\")\n",
    "                    loc.write(\"1 2\\n\")\n",
    "                    loc.write(\"1 2\\n\")\n",
    "                    loc.write(\" {} {}\\n\".format(1 - prev, prev))\n",
    "                    loc.write(\" 1\\n\")\n",
    "                    loc.write(\" {} {} {}\\n\".format(wild_pen, heter_pen, muta_pen))\n",
    "                    loc.write(\"3 {}\\n\".format(gs_num))\n",
    "                    loc.write(' ' + ' '.join(fam_af) + \"\\n\")\n",
    "                    loc.write(\"0 0\\n\")\n",
    "                    loc.write(\"0.0\\n\")\n",
    "                    loc.write(\"1 {} {}\\n\".format(theta_inc, theta_max))\n",
    "\n",
    "            removeEmptyDir('{}/{}'.format(out_base, gene))\n",
    "    tped_fh.close()\n",
    "    tfam_fh.close()\n",
    "    removeEmptyDir('{}'.format(out_base))\n",
    "\n",
    "#parse tfam file, store families into the Pedigree class\n",
    "def parse_tfam(fh):\n",
    "    fams = defaultdict(lambda: Pedigree())\n",
    "    idx = 0\n",
    "    for line in fh:\n",
    "        s = line.strip().split()\n",
    "        fams[s[0]].add_member(s[1:], idx)\n",
    "        idx += 1\n",
    "    return fams\n",
    "\n",
    "#This is to sort the members of a pedigree.\n",
    "#To make sure that parents come before offsprings.\n",
    "class Pedigree:\n",
    "    def __init__(self):\n",
    "        self.fid = None\n",
    "        self.data = {}\n",
    "        self.graph = defaultdict(lambda:[]) #, defaultdict(lambda:[])]\n",
    "        self.sorted = []\n",
    "\n",
    "    def add_member(self, info, idx): #list [pid, father, mother, sex, pheno]\n",
    "        if info[1] != '0' and info[2] != '0':\n",
    "            self.graph[info[1]].append(info[0])\n",
    "            self.graph[info[2]].append(info[0])\n",
    "        self.data[info[0]] = info + [idx]\n",
    "\n",
    "    def get_member_info(self, pid):\n",
    "        return self.data[pid][:-1]\n",
    "\n",
    "    def get_member_idx(self, pid):\n",
    "        return self.data[pid][-1]\n",
    "\n",
    "    def get_member_ids(self):\n",
    "        return self.data.keys()\n",
    "\n",
    "    def print_member(self, pid):\n",
    "        return ' '.join(self.get_member_info(pid))\n",
    "\n",
    "    def get_sorted_ids(self):\n",
    "        if self.sorted:\n",
    "            return self.sorted\n",
    "        else:\n",
    "            #This algorithm was first described by Kahn (1962)\n",
    "            S_no_parents = list(filter(lambda x: True if self.get_member_info(x)[1] == '0' else False, self.get_member_ids()))\n",
    "            graph = self.graph.copy()\n",
    "            while(S_no_parents):\n",
    "                n = S_no_parents.pop()\n",
    "                self.sorted.append(n)\n",
    "                if n not in graph:\n",
    "                    continue\n",
    "                offsprings = graph.pop(n)\n",
    "                for m in offsprings:\n",
    "                    father = self.get_member_info(m)[1]\n",
    "                    mother = self.get_member_info(m)[2]\n",
    "                    if father not in graph and mother not in graph:\n",
    "                        S_no_parents.append(m)\n",
    "            if graph:\n",
    "                raise Exception(\"There is a loop in the pedigree: {}\\n\".format(' '.join(graph.keys())))\n",
    "            else:\n",
    "                return self.sorted\n",
    "\n",
    "#runners\n",
    "def run_linkage(blueprint, theta_inc, theta_max, to_plot = True):\n",
    "    try:\n",
    "        remove_tree(os.path.join(env.outdir, 'heatmap'))\n",
    "    except OSError:\n",
    "        pass\n",
    "    with open(os.path.join(env.tmp_dir, 'LinkageRuntimeError.txt'), 'w') as runtime_err:\n",
    "        workdirs = glob.glob('{}/LINKAGE/{}.chr*'.format(env.tmp_dir, env.output))\n",
    "        parmap(lambda x: linkage_worker(blueprint, x, theta_inc, theta_max, runtime_err, to_plot) , workdirs, env.jobs)\n",
    "\n",
    "def linkage_worker(blueprint, workdir, theta_inc, theta_max, errfile, to_plot = True):\n",
    "    #env.log(\"Start running LINKAGE for {} ...\".format(workdir), flush=True)\n",
    "    #hash genes into genemap\n",
    "    genemap = {}\n",
    "    if blueprint:\n",
    "        with open(blueprint) as f:\n",
    "            for line in f.readlines():\n",
    "                chrID, start, end, gene = line.strip().split()[:4]\n",
    "                genemap[gene] = [chrID, int(start), int(end)]\n",
    "    else:\n",
    "        tped = os.path.join(env.tmp_cache, basename(workdir) + '.tped')\n",
    "        with open(tped) as f:\n",
    "            for line in f.readlines():\n",
    "                items = line.strip().split()[:4]\n",
    "                chrID = items[0]\n",
    "                gene = items[1]\n",
    "                pos = items[3]\n",
    "                genemap[gene] = [chrID, int(pos), int(pos)+1]\n",
    "    with env.lock:\n",
    "        mkpath('{}/heatmap'.format(env.outdir))\n",
    "    lods_fh = open('{}/heatmap/{}.lods'.format(env.outdir, basename(workdir)), 'w')\n",
    "    hlods_fh = open('{}/heatmap/{}.hlods'.format(env.outdir, basename(workdir)), 'w')\n",
    "    famlods_fh = open('{}/heatmap/{}.family-lods'.format(env.outdir, basename(workdir)), 'w')\n",
    "    genes = list(filter(lambda g: g in genemap, map(basename, glob.glob(workdir + '/*'))))\n",
    "    for gene in sorted(genes, key=lambda g: genemap[g]):\n",
    "        lods = {}\n",
    "        hlods = {}\n",
    "        fams = map(basename, filter(isdir, glob.glob('{}/{}/*'.format(workdir, gene))))\n",
    "        for fam in fams:\n",
    "            with cd('{}/{}/{}'.format(workdir, gene, fam)):\n",
    "                units = map(lambda x: re.sub(r'^(\\d+?)\\.PRE$', r'\\1', x) ,glob.glob('*.PRE'))\n",
    "                for unit in units:\n",
    "                    copy_file('{}.LOC'.format(unit), 'datafile.dat')\n",
    "                    copy_file('{}.PRE'.format(unit), 'pedfile.pre')\n",
    "\n",
    "                    step1 = runCommand(['makeped', 'pedfile.pre', 'pedfile.ped', 'n'],\n",
    "                                       show_stderr = False, return_zero = False)\n",
    "                    if step1[1]:\n",
    "                        if env.debug:\n",
    "                            with env.lock:\n",
    "                                print(\"makeped error:\", step1[1], file = sys.stderr)\n",
    "                        with env.makeped_counter.get_lock():\n",
    "                            env.makeped_counter.value += 1\n",
    "                        with env.lock:\n",
    "                            errfile.write(step1[1])\n",
    "                        continue\n",
    "                    step2 = runCommand(['pedcheck', '-p', 'pedfile.ped', '-d', 'datafile.dat', '-c'],\n",
    "                                       show_stderr = False, return_zero = False)\n",
    "                    if step2[1]:\n",
    "                        lines = [x for x in step2[1].split('\\n')\n",
    "                                 if not x.strip().startswith('*') and x.strip()]\n",
    "                        if len(lines) > 0:\n",
    "                            env.log('{} lines'.format(len(lines)))\n",
    "                            with env.lock:\n",
    "                                errfile.write(step2[1])\n",
    "                            if env.debug:\n",
    "                                with env.lock:\n",
    "                                    print(\"pedcheck error:\", '\\n'.join(lines), file = sys.stderr)\n",
    "                    try:\n",
    "                        copy_file('zeroout.dat', 'pedfile.dat')\n",
    "                    except:\n",
    "                        with env.pedcheck_counter.get_lock():\n",
    "                            env.pedcheck_counter.value += 1\n",
    "                        continue\n",
    "                    step3 = runCommand('unknown', show_stderr = False, return_zero = False)\n",
    "                    if step3[1]:\n",
    "                        if env.debug:\n",
    "                            with env.lock:\n",
    "                                print(\"unkn error:\", step3[1], file = sys.stderr)\n",
    "                        with env.unknown_counter.get_lock():\n",
    "                            env.unknown_counter.value += 1\n",
    "                        with env.lock:\n",
    "                            errfile.write(step3[1])\n",
    "                        continue\n",
    "                    step4 = runCommand('mlink', show_stderr = False, return_zero = False)\n",
    "                    if step4[1]:\n",
    "                        if env.debug:\n",
    "                            with env.lock:\n",
    "                                print(\"mlink error:\", step4[1], file = sys.stderr)\n",
    "                        with env.lock:\n",
    "                            errfile.write(step4[1])\n",
    "                    try:\n",
    "                        copy_file('outfile.dat', '{}.out'.format(unit))\n",
    "                    except:\n",
    "                        with env.mlink_counter.get_lock():\n",
    "                            env.mlink_counter.value += 1\n",
    "                        continue\n",
    "                    #clean linkage tmp files\n",
    "                    for f in set(glob.glob('*.dat') + glob.glob('ped*') + ['names.tmp']):\n",
    "                        os.remove(f)\n",
    "                    #collect lod scores of different thelta for the fam\n",
    "                    with open('{}.out'.format(unit)) as out:\n",
    "                        raw = out.read()\n",
    "                        for i in re.finditer(r'^THETAS\\s+(0\\.\\d+)(?:\\n.+?){7}LOD SCORE =\\s+(-?\\d+\\.\\d+)', raw, re.MULTILINE):\n",
    "                            theta, lod = map(float, i.group(1,2))\n",
    "                            #if float(lod) < 1e-6:\n",
    "                            #    lod = 0\n",
    "                            if theta not in lods:\n",
    "                                lods[theta] = {fam: lod}\n",
    "                            elif fam not in lods[theta] or lod > lods[theta][fam]:\n",
    "                                lods[theta][fam] = lod\n",
    "        for theta in sorted(lods.keys()):\n",
    "            lods_fh.write('{} {} {} {}\\n'.format(gene, ' '.join(map(str, genemap[gene])), theta, sum(lods[theta].values())))\n",
    "            for fam in lods[theta]:\n",
    "                famlods_fh.write('{} {} {} {} {}\\n'.format(fam, gene, ' '.join(map(str, genemap[gene])), theta, lods[theta][fam]))\n",
    "            res = minimize_scalar(hlod_fun(lods[theta].values(), -1), bounds=(0,1), method='bounded', options={'xatol':1e-8})\n",
    "            a = res.x\n",
    "            hlods_fh.write('{} {} {} {} {}\\n'.format(gene, ' '.join(map(str, genemap[gene])), a, theta, hlod_fun(lods[theta].values())(a)))\n",
    "        with env.run_counter.get_lock():\n",
    "            env.run_counter.value += 1\n",
    "        if env.run_counter.value % (env.batch * env.jobs) == 0:\n",
    "            env.log('Linkage analysis for {:,d} units completed {{{:.2%}}} ...'.format(env.run_counter.value, float(env.run_counter.value)/env.success_counter.value), flush=True)\n",
    "    lods_fh.close()\n",
    "    hlods_fh.close()\n",
    "    famlods_fh.close()\n",
    "    if to_plot:\n",
    "        heatmap('{}/heatmap/{}.lods'.format(env.outdir, basename(workdir)), theta_inc, theta_max)\n",
    "        heatmap('{}/heatmap/{}.hlods'.format(env.outdir, basename(workdir)), theta_inc, theta_max)\n",
    "    #env.log(\"Finished running LINKAGE for {}.\".format(workdir), flush=True)\n",
    "\n",
    "\n",
    "def hinton(filename, max_weight=None, ax=None):\n",
    "    if ax is None:\n",
    "        ax = plt.gca()\n",
    "    matrix = np.random.rand(20, 20) - 0.5\n",
    "    if not max_weight:\n",
    "        max_weight = 2**np.ceil(np.log(np.abs(matrix).max())/np.log(2))\n",
    "    ax.patch.set_facecolor('gray')\n",
    "    ax.set_aspect('equal', 'box')\n",
    "    ax.xaxis.set_major_locator(plt.NullLocator())\n",
    "    ax.yaxis.set_major_locator(plt.NullLocator())\n",
    "    chrID = re.search(r'\\.chr([0-9XY]+)\\.', filename).group(1)\n",
    "    ax.set_title('Chromosome {}'.format(chrID))\n",
    "    for (x,y),w in np.ndenumerate(matrix):\n",
    "        color = 'white' if w > 0 else 'black'\n",
    "        size = np.sqrt(np.abs(w))\n",
    "        rect = plt.Rectangle([x - size / 2, y - size / 2], size, size,\n",
    "                             facecolor=color, edgecolor=color)\n",
    "        ax.add_patch(rect)\n",
    "    ax.autoscale_view()\n",
    "    ax.invert_yaxis()\n",
    "    plt.savefig(filename)\n",
    "\n",
    "\n",
    "def heatmap(file, theta_inc, theta_max):\n",
    "    #env.log(\"Start ploting heatmap for {} ...\".format(file), flush=True)\n",
    "    if os.path.getsize(file) == 0:\n",
    "        hinton('{}.png'.format(file))\n",
    "        return\n",
    "    lods = []\n",
    "    with open(file, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            theta,lod = line.split()[-2:]\n",
    "            if float(theta) >= theta_max:\n",
    "                continue\n",
    "            lods.append(lod)\n",
    "        if max(lods) == min(lods):\n",
    "            #env.log('Max equals Min for [{}], No real heatmap will be generated.'.format(file))\n",
    "            hinton('{}.png'.format(file))\n",
    "            #return\n",
    "        Num=int(round(theta_max/theta_inc))\n",
    "        lods = np.array(list(map(float,lods))).reshape((-1,Num))\n",
    "        chrID = re.search(r'\\.chr([0-9XY]+)\\.', file).group(1)\n",
    "        fig, ax = plt.subplots(1)\n",
    "        ax.set_title('Chromosome {}'.format(chrID))\n",
    "        ax.pcolormesh(lods.transpose(),cmap=brewer2mpl.get_map('Blues', 'Sequential', 9).mpl_colormap)\n",
    "        fig.savefig('{}.png'.format(file))\n",
    "        #fig.close()\n",
    "    #env.log(\"Finished ploting heatmap for {}.\".format(file), flush=True)\n",
    "\n",
    "def hlod_fun(Li, sign=1):\n",
    "    def _fun(alpha):\n",
    "        return sign * sum(np.log10(alpha*np.power(10, list(Li)) + 1 - alpha))\n",
    "    return _fun\n",
    "\n",
    "def html(theta_inc, theta_max, limit):\n",
    "    if limit <= 0:\n",
    "        return\n",
    "    head = \"\"\"<html>\n",
    "    <head>\n",
    "    <title>Results for {}</title>\"\"\".format(env.output) + \"\"\"\n",
    "    <style>\n",
    "    table {\n",
    "\tfont-family: verdana,arial,sans-serif;\n",
    "\tfont-size:12px;\n",
    "\tcolor:#333333;\n",
    "\tborder-width: 1px;\n",
    "\tborder-color: #CCCCCC;\n",
    "\tborder-collapse: collapse;\n",
    "    }\n",
    "    table th {\n",
    "\tborder-width: 1px;\n",
    "\tpadding: 6px;\n",
    "\tcolor: #006295;\n",
    "\tborder-style: solid;\n",
    "\tborder-color: #CCCCCC;\n",
    "\tbackground-color: #F8F8F8;\n",
    "    }\n",
    "    table td {\n",
    "\tborder-width: 1px;\n",
    "\tpadding: 6px;\n",
    "\tborder-style: solid;\n",
    "\tborder-color: #CCCCCC;\n",
    "\tbackground-color: #ffffff;\n",
    "    }\n",
    "    a,\n",
    "    a:link,\n",
    "    a:visited,\n",
    "    a:active\n",
    "    {\n",
    "\tcolor: #3366CC;\n",
    "\tborder-bottom: 1px dotted #3366CC;\n",
    "\ttext-decoration:none;\n",
    "    }\n",
    "    a:hover\n",
    "    {\n",
    "\tborder-bottom: none;\n",
    "\tcolor: #ffffff;\n",
    "    background-color: #006295;\n",
    "    }\n",
    "    body {font-family:Lucida Sans Unicode,arial,sans-serif;}\n",
    "    </style>\n",
    "    <script type=\"text/javascript\">\n",
    "    function toggle(obj) {\n",
    "    var elstyle = document.getElementById(obj).style;\n",
    "    var text    = document.getElementById(obj + \"tog\");\n",
    "    if (elstyle.display == \\'none\\') {\n",
    "    elstyle.display = \\'block\\';\n",
    "    text.innerHTML = \"hide\";\n",
    "    } else {\n",
    "    elstyle.display = \\'none\\';\n",
    "    text.innerHTML = \"show\";\n",
    "    }\n",
    "    }</script>\n",
    "    </head>\"\"\"\n",
    "    body = \"\"\"<body>\n",
    "    <p><a href=\"#Lods_Table\" onclick=\"toggle(\\'lods_tbl\\')\">Ranked LOD scores</a>\n",
    "    <div id=\"lods_tbl\" class=\"divinfo\", style=\"border:0px;width:auto;height:auto;overflow-y:hidden;overflow-x:scroll;\">{}</div></p>\n",
    "    <p><a href=\"#Hlods_Table\" onclick=\"toggle(\\'hlods_tbl\\')\">Ranked HLOD scores</a>\n",
    "    <div id=\"hlods_tbl\" class=\"divinfo\", style=\"border:0px;width:auto;height:auto;overflow-y:hidden;overflow-x:scroll;\">{}</div></p>\n",
    "    <p><a href=\"#Lods_Heatmap\" onclick=\"toggle(\\'lods_heatmap\\')\">LOD scores heatmap</a>\n",
    "    <div id=\"lods_heatmap\">{}</div></p>\n",
    "    <p><a href=\"#Hlods_Heatmap\" onclick=\"toggle(\\'hlods_heatmap\\')\">HLOD scores heatmap</a>\n",
    "    <div id=\"hlods_heatmap\">{}</div></p>\n",
    "    </body>\n",
    "    </html>\"\"\"\n",
    "    env.log('Generating Report in HTML format ...', flush = True)\n",
    "    with open('{}/{}_Report.html'.format(env.outdir, env.output), 'w') as f:\n",
    "        #t = Template(index)\n",
    "        #c = Context({ \"lods\": lods_tbl })\n",
    "        f.write(head + body.format(html_table('Lod', theta_inc, theta_max, limit), html_table('Hlod', theta_inc, theta_max, limit), html_img('lod'), html_img('hlod')))\n",
    "    env.log('Report for [{}] is generated in HTML format\\n'.format(env.output), flush = True)\n",
    "\n",
    "def html_img(ltype):\n",
    "    chrs = ['{}'.format(i+1) for i in range(22)] + ['X', 'Y']\n",
    "    imgs = ['heatmap/{0}.chr{1}.{2}s.png'.format(env.output, chrID, ltype) for chrID in chrs]\n",
    "    exist_imgs = [img for img in imgs if isfile('{}/{}'.format(env.outdir,img))]\n",
    "    return ''.join('<img src={}></img>'.format(img) for img in exist_imgs)\n",
    "\n",
    "def html_table(type, theta_inc, theta_max, limit):\n",
    "    colNum = int(round(theta_max/theta_inc))\n",
    "    theta_orders = range(colNum)\n",
    "    thetas = map(lambda x: theta_inc * x, range(colNum))\n",
    "    #table\n",
    "    table = r'<table style=\"width:300px;font-size:12px\">{}</table>'\n",
    "    #table header\n",
    "    lods_header = r'<tr>{}</tr>'.format(''.join(r'<th colspan=\"2\">&#952;={}</td>'.format(x) for x in thetas))\n",
    "    lods_header += r'<tr>{}</tr>'.format(r'<th rowspan=\"2\">{}</th><th>Marker name</th>'.format(type) * colNum)\n",
    "    lods_header += r'<tr>{}</tr>'.format(r'<th>chr:start-end</th>' * colNum)\n",
    "    #initialize lods dict\n",
    "    lods = {}\n",
    "    for theta_order in theta_orders:\n",
    "        lods[theta_order] = {}\n",
    "        #print '{}\\n'.format(theta_order)\n",
    "    #read lods\n",
    "    lods_files = glob.glob('{}/heatmap/{}.*.{}s'.format(env.outdir, env.output, type.lower()))\n",
    "    for file in lods_files:\n",
    "        with open(file, 'r') as f:\n",
    "            for line in f:\n",
    "                if type == 'Lod':\n",
    "                    gene, chrId, start, end, theta, lod = line.strip().split()\n",
    "                    if int(round(float(theta)/theta_inc)) >= colNum:\n",
    "                        continue\n",
    "                    lods[int(round(float(theta)/theta_inc))][gene] = ['<b>{}</b>'.format(round(float(lod),3)), gene, chrId, start, end]\n",
    "                elif type == 'Hlod':\n",
    "                    gene, chrId, start, end, a, theta, lod = line.strip().split()\n",
    "                    if int(round(float(theta)/theta_inc)) >= colNum:\n",
    "                        continue\n",
    "                    lods[int(round(float(theta)/theta_inc))][gene] = ['<b>{}</b><br>&#945;={}'.format(round(float(lod),3), round(float(a),3)), gene, chrId, start, end]\n",
    "                else:\n",
    "                    env.error('Wrong type of LOD')\n",
    "\n",
    "    #collect result\n",
    "    res = np.empty((len(theta_orders), limit), dtype=list)\n",
    "    for theta_order in theta_orders:\n",
    "        i=0\n",
    "        for gene in sorted(lods[theta_order].keys(), key=lambda x: lods[theta_order][x][0], reverse=True):\n",
    "            if i >= limit:\n",
    "                break\n",
    "            res[theta_order][i] = lods[theta_order][gene]\n",
    "            i += 1\n",
    "    #print res\n",
    "    #write lods table\n",
    "    lods_res = ''\n",
    "    for i in range(min(limit, len(lods[0].keys()))):\n",
    "        lods_res += r'<tr>{}</tr>'.format(''.join(r'<td rowspan=\"2\">{}</td><td>{}</td>'.format(res[theta_order][i][0],res[theta_order][i][1]) for theta_order in theta_orders))\n",
    "        lods_res += r'<tr>{}</tr>'.format(''.join(r'<td>{}:{}-{}</td>'.format(res[theta_order][i][2],res[theta_order][i][3],res[theta_order][i][4]) for theta_order in theta_orders))\n",
    "    lods_tbl = table.format(lods_header + lods_res)\n",
    "    return lods_tbl\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testing runner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SEQLinkage.Main import *\n",
    "from __future__ import print_function\n",
    "from SEQLinkage.Utils import *\n",
    "from multiprocessing import Process, Queue\n",
    "from collections import OrderedDict\n",
    "import itertools\n",
    "from copy import deepcopy\n",
    "import sys, faulthandler, platform\n",
    "import numpy as np\n",
    "import os\n",
    "if sys.version_info.major == 2:\n",
    "    from cstatgen import cstatgen_py2 as cstatgen\n",
    "    from cstatgen.egglib import Align\n",
    "else:\n",
    "    from cstatgen import cstatgen_py3 as cstatgen\n",
    "    import egglib\n",
    "    from egglib import Align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "args = Args().parser.parse_args('--fam data/mwe_normal_fam.csv --vcf data/first1000snp_full_samples.vcf.gz -f MERLIN --blueprint data/genemap.hg38.txt --freq AF --run-linkage -K 0.001 --moi AD -W 0 -M 1 --theta-max 0.5 --theta-inc 0.05'.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Namespace(bin=0.8, blueprint='data/genemap.hg38.txt', single_markers=False, tfam='data/mwe_normal_fam.csv', vcf='data/first1000snp_full_samples.vcf.gz', build='hg19', prephased=False, freq='AF', freq_by_fam=None, mle=False, rvhaplo=False, recomb_max=1, recomb_cross_fam=False, rsq=0.0, include_vars=None, maf_cutoff=1.0, chr_prefix=None, output=None, format=['MERLIN'], prevalence=0.001, inherit_mode='AD', wild_pen=0.0, muta_pen=1.0, theta_max=0.5, theta_inc=0.05, run_linkage=True, output_limit=10, jobs=16, tempdir=None, vanilla=True, quiet=False, debug=False, no_save=False, func=<function main at 0x2b7a3363e1f0>)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1;40;32mMESSAGE: Binary trait detected in [/mnt/mfs/statgen/yin/Github/linkage/SEQpy3/data/mwe_normal_fam.csv]\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checkParams(args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if args.no_save:\n",
    "    cache = NoCache()\n",
    "else:\n",
    "    cache = Cache(env.cache_dir, env.output, vars(args))\n",
    "cache.setID('vcf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Testing run_linkage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args.run_linkage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cache.setID('analysis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'LINKAGE'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\u001b[0;31mSignature:\u001b[0m \u001b[0mcache\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtarget_dir\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnames\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
       "\u001b[0;31mDocstring:\u001b[0m <no docstring>\n",
       "\u001b[0;31mFile:\u001b[0m      /mnt/mfs/statgen/yin/Github/linkage/SEQpy3/SEQLinkage/Utils.py\n",
       "\u001b[0;31mType:\u001b[0m      method\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?cache.load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/mfs/statgen/yin/Github/linkage/SEQpy3/cache/LINKAGE.cache'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cache.cache_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not args.vanilla and cache.check()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fmt = args.format[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/genemap.hg38.txt'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args.blueprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.05"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args.theta_inc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args.theta_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args.output_limit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\u001b[0;31mSignature:\u001b[0m \u001b[0mrun_linkage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblueprint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtheta_inc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtheta_max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mto_plot\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
       "\u001b[0;31mDocstring:\u001b[0m <no docstring>\n",
       "\u001b[0;31mSource:\u001b[0m   \n",
       "\u001b[0;32mdef\u001b[0m \u001b[0mrun_linkage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblueprint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtheta_inc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtheta_max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mto_plot\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m        \u001b[0mremove_tree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0menv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutdir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'heatmap'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0;32mexcept\u001b[0m \u001b[0mOSError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m        \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0menv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtmp_dir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LinkageRuntimeError.txt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'w'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mruntime_err\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m        \u001b[0mworkdirs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mglob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mglob\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'{}/LINKAGE/{}.chr*'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0menv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtmp_dir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0menv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutput\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m        \u001b[0mparmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mlinkage_worker\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblueprint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtheta_inc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtheta_max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mruntime_err\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mto_plot\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m,\u001b[0m \u001b[0mworkdirs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0menv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjobs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
       "\u001b[0;31mFile:\u001b[0m      /tmp/2206534.1.plot.q/ipykernel_30528/2972776299.py\n",
       "\u001b[0;31mType:\u001b[0m      function\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "??run_linkage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LINKAGE/LINKAGE$ head LINKAGE.chr1/DDX11L1/1036/*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "env.tmp_dir = './LINKAGE'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "env.jobs=8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blueprint, theta_inc, theta_max, to_plot = args.blueprint, args.theta_inc, args.theta_max, True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "--fam data/mwe_normal_fam.csv --vcf data/first1000snp_full_samples.vcf.gz -f LINKAGE --blueprint data/genemap.hg38.txt --freq AF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./LINKAGE'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env.tmp_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'LINKAGE'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['./LINKAGE/LINKAGE/LINKAGE.chr1']\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAOsAAAD7CAYAAACL3GNOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAAd20lEQVR4nO2dX4xlSV3Hv7+FXdbtRkfDzr09qEjsSNAnExJfeNiIRsGoPAtPJsosxmjsDBjDOL2M+G8zyRrjuvPgi1FQ/B8fjIYYHwjyIIaQGIO5JEv4c28PYjD0uAjSx4e+p6murjr3W6eqblWd+/skk+k+p7rOqTr1O/ecqu/9/qTrOiiKUj+PlD4BRVE4NFgVpRE0WBWlETRYFaURNFgVpRE0WBWlETRYEyIixyLyR6XPQ5kmGqyBiMhPici/iMipiCxF5O9E5I2lz6s1ROQxEflzEXlRRDoRear0OdWOBmsAIvJLAJ4D8OsAZgC+E8DzAH5yRF0vT3pybfJhAG8HsCp9Ii2gwUoiIt8C4L0Afq7rur/suu5h13Vf67rub7uuu2UUfUxE/lBEviwi/yYibzDqeFFE3i0inwDwUEReLiI/sS73JRH5JxF5vVX+loh8QkQeisgfiMhs/Wn+ZRH5kIh8q1F+qK53i8jn1n/3SRF503r7K0TkORH5/PrfcyLyivW+p0TksyLyLhF5sH6SeKuIvEVE/kNE/ktEfsU4xiMi8ssi8ikR+aKIfFBEvs3Vn13XfbXruue6rvswgK9HX6BdoOs6/Uf8A/CjAP4PwMsHyhwD+AqAtwB4GYDfAPBRY/+LAD4O4DsAfBOA7wHwEMAPA3gUwLsALAA8ZpT/KM4/xV8N4AGAfwXw/QAeB/CPAO6sy3rrAvA6AJ8BcGNd9rsAfPf65/euj3EdwJMAPgLg7nrfU+s2/+q6zp8B8AUA7wfwSgDfB+AlAK9dl/+FdV3fDuAVAO4D+ADRt58F8FTpa1z7v+In0Mo/AG8DsNpQ5hjAh4zfvxfAS8bvLwL4aeP32wA+aPz+CIDP9QN3Xf5txv6/APD7xu8/D+CvN9UF4HAd6D8E4FHrnD8F4C3G7z8C4MX1z0+tg/Fl699fCaAD8ANG+Y8BeOv6538H8CZj3wGAr2HgBrcup8FK/NPHYJ4vAngV8a5pvn/9D4DHrb/5jPHzDQCf7n/puu5svf/VRpkT4+eXHL/vb6qr67oFgF/E+c3kgYj8iYjccP3d+ucbxu9f7Lquf0x9yXNO/Tm8BsBfrR/Dv4Tz4P06zp8MlEg0WHn+GcD/AnhrZD3m15w+j/MBDgAQEcH5I/LnRtQ7WFfXde/vuu6N6zIdgN9y/R3OJ80+P+L4wPnN4c1d110z/j3edd2Y9igWGqwkXdf9N87f3X5vPcnyhIg8KiJvFpHfHlntBwH8mIi8SUQeBXCE8xvCR1LWJSKvE5EfXE8cfQXnn4Zn67/7AID3iMiTIvKqdRvHrhW/AOB9IvIaAFjX6Z0pX09uPb7+9TEReXx9k1Ec6PJBAF3X3RORFYD3APhjAF/G+Tvb+0bW90kReTuA38X5o+/HAfx413VfTVnXOkh/E8Drcf4O+REAP7v+018D8M0APrH+/c/W28bwOwAEwD+sH7MfAPhTAH/jKf9JfONT/e/X/78W5+/qioWsX/AVRakcfQxWlEbQYFWURtBgVZRG0GBVlEbQYFWURghaunniiSe6a9euZToVRVGWy+V/dl33pGtfULBeu3YN73jHO9KclaIoVzg+Pv60b58+BitKI2iwKkojaLAqSiNosCpKI2iwKkojaLAqSiNosCpKI2iwKkojNPHl86OjI+zv7zv3nZ6e4t69ewCAZ599Fg8fPrxSZm9vD7dufcMtlC3H8vzzz+PBgwdXts9mMywWi4vzY/Gdn32OvuPa5Xz9Z/bdUDlX2W214/r163jnO98ZdFy2Pra9Q+XOzs5w9+7doPMbS/FgtTvCNSh8HWXv8w0MeztbjsU3ME5OTq6cuz2QXINx6DzMfb7j2uV8/WdvZ/sZ4K5binYM7RvzN+a+/vzn8zlOTs494GazGVar1aW2DfXLI49cfjh13aDGfgjYZA3W/sSHTvbw8PCio4Dzznr66adznlZR7IE0ZjCOxTUox2IOdABRdYVy8+ZNzGbnhoknJyd44YUXouozx5/58xj6QF0ul5jP51itVjg4OIiqsydrsPYnPnSHtTsntrNy0H+KjHkUrImUgzJ1PSH0gWr/XBP9Taz/PwU6waREMZvNqg2YkvRPGimfOLJ+su7t7V08Bvu4fv36lXe42kj5aVqyvbPZ7NJjcAynp6fY39+/GIynp6fR58dycnJy6TE4lpT90pPq0dcka7AyL9WhM31D9DcH1/Yx5VjMi21vt0nZXhY7sMztY2FuYL5+7vf1+Pqv32cT+45qk/LTz9XmsePKpvhsMEM/2Hz7etgZtxQzcyaLxYI6PxZ2kLPl2CcDtp9Z2H7e1H+hTzb204u9z6ybaW9Iv6QeWyZBvsE3btzo9MvnipKP4+Pjj3Vd9wbXPp1gUpRGyPIYbK6DmYxdE2MVR77jxhybgVUIpYbtl9QKsNhzHHt+qY8bwu3bt68IIOzzy9EOkyzB6guYsTNtrOJoqH57n0uSxg4iu+NZhRDgvuiuC2mXc8na2H4JUYD1i/kmq9UK9+/f99YxhOscx55f6uMC7mvsuh6uQLXPL0c7TKqYYCohOnBNQLCDKKbjXRfdVZ9dzjdYUuNaxPct7KdWErGYARY7Zlx9nyKwclDFO+v+/j6Oj4+r7STFTSklEftpNjWqCNZ++nubC+uK0hpVBOu9e/cgIk3rblNzdnY2+HsN5NAaM9hroLtCFe+sm0itOPLVadfnWwy3BwhbzlfWVY75jmTqftnb28NqtbryjuoLxFTvqKFijLHqKVe/sNfj7OzMOxu8LbKIIkotZZRauqmd3EsKyjkp+nlIFKEKJkWpCFUwKcoEKKpg8nnl2FYnqR/jYj2J7HI+dctYfx62vexjP1ufrx12OdZbqXZStyN3vxRVMPm+GWFvT+0NFOtJZG/3DXDGnwe4eiHZ9rKKLba+IeEF43Xl2scqhFw3CtfNjq2PIaQdvmt3fHw8qr4xNDEbDFz2DwL8HkL7+/sXZWN9huxjx9aX2qitBViFkOtGwaq9fDcjO8DMwAol5BqZks2UHkzNvLOGeDX1+1L7DKVcS1wul1gul8nqq535fJ7Uj4ih1E3QbKd6MG2gfwRUbyBlSjQTrHbgDQXiarVC13XJ7DpyBP/BwUEWn55aWa1WW7UrBa66amwLs53VG6a5VDD99pg6GWxFSozChPUu8qlbbIlgDiVW7bAKodTlgLQWKyHeSrluwlmC9fDwkJr1ZAdviByNmRVk5YHsDCO7PMMOHra9psufjfl+ndpriPV+Avg+TF2OIaQdzLULqW8MqmBSlIpQBZOiTIAsj8GsoqeU4L8UQ+t85r5YzySgDiVR7ecH8F5NqVVvY8gSrKyiJ7VtSojhlytw7HKsHNJ37LEDMjZr3qZ92yKH0sl3wxsreGC9mlKr3sbQjILp6OjoIuOcL+9prELILsfKIX3H2FbAsIoZ8+YzdCMJ8Tg6OjoCED/xYyrPgPgsdz1MJsNWaCZY9/f3LymJdsl7ZxOsYsa8yQzdSEI8jlJehxzOE0wmw1ZoJliVOqkxRaeLKaxlNxOsp6enFwmMZrPZTnnvbMIUoQw9OprrgEOD11xv3dTPKd03cmRzixHv10YzwXrv3r1LGdFd70ixCiG7nC/BkStNY87sYZtgFTPsO1uJmfjT09OkWe5Sk8OPK5QswcrK71I3LGQCgbnjhqRnZI7N3uVzKIlKkEPplPqTkq2vhMLKRhVMilIRqmBSlAmQ5TGYFRMMPYKY62Kx2dJyq0xYpU7t2eZY2Haw/VK70qkWK9cswRoiJvBhXrzYbGkxKhPblMzlP+zKvtaLE8xztBf+e2az2aXJM9dxV6sVDg8PN8rgXIMnRCzCKInY/mMVTDmUTiUUZblpZja4FOyX3m0xgk+c4FqXdG2zjzOfzykZXOzgqT2rGnt+IYoyM7CHAppVWKX0fjJp7p31zp07pU9BmRjspzyrsMqllmouWBVlV2kuWJ955pmtHs9+1NmWjxB7XHtNuhX531h8Vi8xsF5NbD7aXGvc2Uy+mfcwllj7l5iLaVvU+N5pbN8pV3CZkkkTlyLKZY3jGgRjM6r5YDyOfP28jbSU7MxriKKMnXRib9S5Zq+zBOtisYheojA7NrVMLiSoWe8dn+TPbIctmRwi5QUPqYvpw5iAMfeFlgshZf+FpqTMhSqYFKUiVMGkKBOgaBa5Usqa1PXFKpjsOmtRzPhI3d4Uxx4zZnJ4XTXnwcRmkUudqClEmVRCqcMqYdhysZ5TQ9uHSN1egFccsWPGJXm15a45vK4m7cHkk+m56AMsxV2qdqUOQ66sdKyiJyW2ZDM2+5pL2hoid62R4u+srEwPOF/OEBEcHh7mPq2dZrFYoOs6LJfLrXoX5cq+NhWKB6uiKBzFgzVEIdTf8ReLRe7T2mn6J5iDg4NJZF+bCsXfWUPeS1LOgoZkI6uVXFnpSnx/1CUscbWDbbPLP8ulFGuJLMHqy25my+xSD7YQZVLKbHOsAid1NrfUnlMsqdsL8G1hyzH+WTm8rnIapqmCSVEqQhVMijIBir6zxnomucqmPG6r3kClPJ2AvEoinzIOuKyOK6UUyz1esgTrs88+i8VicWWtzJYbxnom2fvYwcse19UG4OqCfUjQMMqp1O0AuKx5AKckClH0pFQSDX3F0tyXWikGXE6M1mO3I3dWvyyPwQ8fPnQO8lQpEXyEDN75fA4RufjnOl/fwry9PeS4zLackjUT1qcoZqBNRUlkJkbr2XY7dvad1e74qTsstMJ8Plf1koedDVZFaY3ioohS2PYquR/RFQ5VLvnJ8sm6t7fn7PTcj5q+hWfX9tVqha7rLv65ztc3cNhsZ6y5l8vjiKnPV471QvIphJhtLC7VUItKot4/y2Tb7cjyyXrr1i3cv39/Y7nYNHp22dQeTC7Tsh7bW4klpcdRyHFZBRPrOcUqelIqiXzKuH5faH0hCivGPyt3Vj9VMClKRaiCSVEmwKQmmEopjlhlze3bt51JpoFxnj9sFr5SSp1Nj979fl/WQeCy8ICtj6V2hZpNtmC1B+bZ2Rnu3r17pZzdYa5OYhU9qRUkrMcRq6zxBSpw1fPHtrsBwqxOxnohMYqj1P08JC4YKzxgxkxIP7MeUa4btyvz4BiyBas9MH0D1e4AJlPapu0MzE0il8cRQ4iiiq3PXqpyzXZPSXHkyvpmjxm2n1lll+vGnWpZsIp31uVyeeH5sy2Ym8SU2EXFFpv1rRWqCFZFUTZTRbAeHBxceP5sC3vda5teQyVgk0JPCTbrWytke2c9Ozu7MsHkwl5IdgVNDquMmMX/bQS2nZWu3xZTH8NUvItOT0+9mfxM2H5ms9K5hBupHsGzBatr5tcFEzSps5axsFP3rLLGvoGZ2J4/TFa6IcZ6ITGKo9T97LpBmPvGwKbCZPuZHQspZn19qIJJUSpCFUyKMgGy2bqkzA7HEpu9buw5pm7vtuqLqZMhR5a2XSZLsLJigljjMlZJxGav27Qv9G/Grt1uqz7XPkYsAnBKnRxZ2hgPK1c7euz2+KSOTCpH13F9ktJqUz6yhKRUDMk2lxL7otvSu9j6UnxymAMpdlCwYpGcSp0hWF8r9obnm9hiUjm6tvkmEFP4ZzXzzhqSbS4l7MUdW18K5VSfm2Y+nzeXtlLhaSZYFWXXaSZYQ7LN1UwO5VSfXW+1WjWXXEvhKfrOGqJM2qYU0YRRWIWQY2Yzpfs+296cSp0hXGOG9ZxKfdxt3xizBCsr04tVJtn1xWavc9VZKv1hSpljjvYySp0cWdpSj5nUY1WzyCnKjqAKJkWZAEUVTGw5VjzBehylJra9rrIlKJUdLrVHVGpC+iVnZr+iCia2HCueYD2OfB3q8spxKVxcWdV8Xj6mf3KOLGOsN1DK7HDszSlHNjeWlJ5JIdfNzjTXM5vNNvoOb2JS7oYsvgHgmpxyiSBcFy+1ZxJw1a3P5d4XkwmuFjlkKGy/uFRvtvl8aiWWb0Y8xUy5vrPuECW8rkpSSvWWCw3WHaIfrK0P2l1Fg3WH6FVfraq/dp2dfGf1LVy73itcliMucUJqzySAc5hnvYFcFiatmsSx/WJfE9f1LaXEGkNRBRMLqwphPY5CptBZTyKXHNKefcyRZYxdTkmZHY69vjmyubGwmQwZJVbIdfP5SaUwnVMFk6JUhCqYFGUCNOHB5LPemM1mWCwWwcqQ1CoTVuHCKqxKKbFqVxLVfn65KapgAjibE587w8nJyShlEquIYg3YWIULq7BiywHuyZaxN0U7mVOPK4kVo/xhU2GGnB+7j1UwpfZ0yknx2eCUNichyiQG1oCtNpibIuAOdN+MqQ2j/GFTYfaYTxS+FKEAnNnhbFjFlusG5ZIGllZtAfrOujO0kCXPfKIYerpInR0uR4a9mzdv4ubNm9H1mGiwKkojFH8MVpRQZrPZpcfgGsmR80Y/WQfwKZBql+v5FEybygD8t1DYd9sQbKGEj9Vqha7rLkziXDB90B/Hbl+tWfOKK5gY6ZhPFTKbzSgljH3h2XKHh4fUDCCrcGEVVmw5gOs/gFMw+dIkuo7LfHKw2fV6Ni29hCid2Bnae/fuUd8zLZn+s0cVTIpSEapgUpQJMKkJppz+N0PUsGA+xK4rf6ZCtmBlEzC53rlc2xilCatMYusD3NI/e8E+x4I5o8BJ7YUExCl6YhRCLCnPL8dNzCWNjU1m1pMtWFMnYGKzeKWuzzXRM7Rgvwn2JmZ7CLmy5uW4SbD9slgsNhrEhdR3dHR0yWzMp/uOySLHGvH59jEKK9dEaGwys57m3lnn83nTtiQhNzGznbW1ObVB3P7+/hVlUm0Z8ViFVbbjb/2IiqKMorlgXa1W1YsSUmG2c1faXDNmAqxtJMOyyfbOmjr7WuosXi7hgas+V7mYC8X2SynPJLafWe8itr5eSWS+s8ZcX9abKgTft4BMXAKeVIqoSYkidm3pJjb9CNDe0k3K2eUa+2VIFDGpYFWU1lEFk6JMgJ1QMAGXH2tqVxyxlMpKl/rxsVQ7as/qZ9NEsA59s8Tcxy5yh4gJGI8j9l055CaRMusbWx9bju1nNqjZdsTejIHLbYntP7s+3zmmev+tIljNjqjxjmYzNiVlyE2CUTCFUCLbXOpUjrE34037hghJT8p4RI0h6zvr0dER7ty5g6Ojo8FyDx8+xJ07dy5+Vs6pWcGk+EntEdWTNVj7u11tsjFFaZGswdovam9a3N7b28Mzzzxz8bNyjiqY2sQUi6T0iMr6zsq+VNf+jmqzjRtKagVTjKJHb6A8LmucGCWWSRUTTKlgvYtSe0Sxnk4hx02Z9Y2tD4hrb78vtBwLW19sNjzgqkSQvXY5VU+qYFKUilAFk6JMgEk9Btcm5AfGrRvXrqxhxQlsuaGseaYjQ+rjstQi+G8iWNkgDPFgYmFsWFIriULqYzyiUsOKE9hyQ64L5r6Q47LZ8FIqtgD/O3+Km2xxwzQ7EF13qhxByJLaSyqlQgjgPaLs9ItDKRdbU5S5YDMGpFaK+Ugh9ilumGYH3DYFFP2nUu5PohoIScfYD2AAWQZubbSiFGviMTgX/SdQCfOr2ql50O4q2UYpmwjJlYdmW/T2LCX8dGpmb2/vwutKBRH1kO2TNSQxUCmYR9/UXlK+Y4yF9YiyPZOGBOa3bt268P9t8X0VuJwW0txm09+YzHfWWm9QTTwGp84OF0JqJRGjEAqpj33XTpkvNLWCaUh5Zt54Qo7LZsMzb0zmNpOQbHg5s82pgklRKkIVTIoyAZp4DJ4KtStrpqLEql0BNpatiSKAqx3lyrgFpMu6NYQtEgCuCgVY3x2fXM5evy1lc8LCKqd8/WJfN1tw0GMLD1LfdHJ4U7nGqi95Vi62JopwbfNl13JtZ5ROIbgmDOxtrO+Ob3Ikdv22ViWRr19c141JYBV60zEDJ/bGzirKXG3zJc9KPVZ7mnkMDlE6mTOuzOxrrZiDJoVcrX+aGJIatoAZOKnSKabE1CY3Y5im1EX/5JDSakRx0y/pNGOYpsRhuhWkWKfLMYBKkMvjKCU5boxbyyLXbxuLvSA+JHRgHn3ZLGglST3JVuLR11QHmdtiWCwWF2Mh9n2QHaesIqo/J7ON1XswMZMhrvR4/Xab1DNuzMBl1Sg+BY4t+yvlSZQCxrvIpQf3fWvHLBvajk1jIYc3lXmDiD2/saiCSVEqQhVMijIBii7dpE40VIpaPHq2RWovpNqvL0vucZAlWFllEuuVE5tlzO4oNisdS4jnD4OvHWPXR9lBNNR2M2jY9pZMJMV6ITFjhu2/1IoymyzBGqJMYr1yGHJ4Ndk3Hp9ihm1HjOeUb/bRlE66AjrFILKDJuV1C8UOxJBP3zEZAHMHIYu+s27AvsHEKmb6TyURwXw+T3KxQ9Yd5/P55CxbdiXzoAZrAaYiTlC2SzPa4CnRL7CXUN9oNrp2yfLJ6huEzDddhv6+BLZAwyXYAPh29OqWruuwWq2SiBdsS5NtUPK6sWZ8UyPLJ+uQ2sOcTGG9clhFSo5sbozkL8Tzh5m+97XD99i8qc4USie7n5n2ssf1Kdn6fTYxvlj2NWbGTKmseTaqYFKUilAFk6JMgCYmmKaicGHZNa8mlk2ClVBBC9uOWpRY2YKVtbZImVWtpKcTw9DFNAcauwifWlnD1pfD46gEi8XCu+a8Wq0u/IRLKrFMsgUra8Nid1hMFq8YTyeflM82Q/MlsbLrc5VLvXhfuwEbwHsc+T4VY2x5Nnk1DYlDahSOFH9ntTtlW51kDz7fsoP9PVWfCZpdnya7Kk/tXk2h6IhSlEbQYFUmi7lG6xOztES2d1bWM8mW3W0rixd7frZliy89ZIhHlLIdaphUTEm2YGWXA1yTOvZMIatgyuHpxGZpY+qLVdX0+0qXC/E4Yg3JSvg7s9nh2HIh/TIGVTApSkWogklRJkATCqapULsSq5SX1K55WI0lS7AeHR3h8PDw0vO8K+OWK5MbcFWgwJarnVgvKWCcvI31popVOvkEI5sIEWOwmf0YhvoPGPcenfL8bLIE6/7+/pWvc7kybrHfew35fmxIBi+zLCuH9HW668KOnTRJLW8bMjh7+umnk52fSwjiCmxXX7OGc2xmv1gZq4tNXld9nWaay16Rl0K9Nrl31pBsc4wplt3Jrfr9lDI4c/Ur+13dofNbLpdYLpfe/f11MsvFXjvW68pU4aVU5Ok7qzJpatT4jmWng9VcP1MRQ1uwX/Yo4TllJuNKefxswWpn3dqWP0+IkmhMcqpW/X5CsqClxCW0cF2T1OfXXzczqGOvndmWoXE19ltjm8gSrD6PHruBPmWIa3KKKQekz+DFzuAxk0mxXlL9vjH1+a6H7YkVo3Qam1EtxMOKVYHF+DT5SKlSG4MqmBSlIlTBpCgTYKcnmFpnKt5KtVNLv2QJVlbFwZZjFTipz89XNub8GJFAjqxlKb2u2HaElEvp1ZRaAZajX8aQJVhZlQlbLiQ7HKMkYo/LbuvPw1Th+NI9MiIBl+Kory+3FxILK3Zgt4Wc3yblWUmDM7a9Y8j6zrpcLtF13aDSZErY34FMVVeK+qYEm45xamQN1n5heEoqEkUpRdZg7R8DNXOZosSTNVgPDg4gItkUHbURktQ4pK4U9bmsbVpVY5XImlcDWSaYWBVHykxfPTFKItfgZTyE+vOznyBc58fI71JnaQM487AQDyFWRsiWY72agHRZ89j25uiXMaiCSVEqQhVMijIBVMGkTI5aFEepKapgCiGlMoRVHLFKmNQKK182PJePFYPvetjJmkLaa3ts+c4vpdIpteIoh1Gbb86kWg8m24cG8GeHswemLz0jqwxhPJhYRVSsJQxzfq5z9CVRcvlYMfgGr32ckPb6bFiYPhy7LbXiKDRrnnmzCA2+qj2Y2Oxw9oCJzfYVEnAlyCVHOz4+LuJqv0uYAVfCi0snmBSlEXSCaSK0KnBQeLIFq2ka1f/uwk4m5UvNZ2dzA7hF+NoULrkWzVud4WwJUxxR4uaYTcHkmkxyNZBNy5cymxuriCrlSeTLhjebzUYFtk+BY98YQ9rrMzhj+nCs0im14ihEAQbE+TqpB5OiTAxVMCnKBNBgVZRG0GBVlEbQYFWURtBgVZRG0GBVlEbQYFWURtBgVZRGCBJFiMgXAHw63+koys7zmq7rnnTtCApWRVHKoY/BitIIGqyK0ggarIrSCBqsitIIGqyK0ggarIrSCBqsitIIGqyK0ggarIrSCP8PUQ8JZNbEoCoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with open(os.path.join(env.tmp_dir, 'LinkageRuntimeError.txt'), 'w') as runtime_err:\n",
    "    workdirs = glob.glob('{}/LINKAGE/{}.chr*'.format(env.tmp_dir, env.output))\n",
    "    print(workdirs)\n",
    "    linkage_worker(blueprint, workdirs[0], theta_inc, theta_max, runtime_err, to_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = {'a':1,'b':2}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_values([1, 2])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp.values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_linkage(blueprint, theta_inc, theta_max, to_plot = True):\n",
    "    try:\n",
    "        remove_tree(os.path.join(env.outdir, 'heatmap'))\n",
    "    except OSError:\n",
    "        pass\n",
    "    with open(os.path.join(env.tmp_dir, 'LinkageRuntimeError.txt'), 'w') as runtime_err:\n",
    "        workdirs = glob.glob('{}/LINKAGE/{}.chr*'.format(env.tmp_dir, env.output))\n",
    "        parmap(lambda x: linkage_worker(blueprint, x, theta_inc, theta_max, runtime_err, to_plot) , workdirs, env.jobs)\n",
    "\n",
    "def linkage_worker(blueprint, workdir, theta_inc, theta_max, errfile, to_plot = True):\n",
    "    #env.log(\"Start running LINKAGE for {} ...\".format(workdir), flush=True)\n",
    "    #hash genes into genemap\n",
    "    genemap = {}\n",
    "    if blueprint:\n",
    "        with open(blueprint) as f:\n",
    "            for line in f.readlines():\n",
    "                chrID, start, end, gene = line.strip().split()[:4]\n",
    "                genemap[gene] = [chrID, int(start), int(end)]\n",
    "    else:\n",
    "        tped = os.path.join(env.tmp_cache, basename(workdir) + '.tped')\n",
    "        with open(tped) as f:\n",
    "            for line in f.readlines():\n",
    "                items = line.strip().split()[:4]\n",
    "                chrID = items[0]\n",
    "                gene = items[1]\n",
    "                pos = items[3]\n",
    "                genemap[gene] = [chrID, int(pos), int(pos)+1]\n",
    "    with env.lock:\n",
    "        mkpath('{}/heatmap'.format(env.outdir))\n",
    "    lods_fh = open('{}/heatmap/{}.lods'.format(env.outdir, basename(workdir)), 'w')\n",
    "    hlods_fh = open('{}/heatmap/{}.hlods'.format(env.outdir, basename(workdir)), 'w')\n",
    "    famlods_fh = open('{}/heatmap/{}.family-lods'.format(env.outdir, basename(workdir)), 'w')\n",
    "    genes = list(filter(lambda g: g in genemap, map(basename, glob.glob(workdir + '/*'))))\n",
    "    for gene in sorted(genes, key=lambda g: genemap[g]):\n",
    "        lods = {}\n",
    "        hlods = {}\n",
    "        fams = map(basename, filter(isdir, glob.glob('{}/{}/*'.format(workdir, gene))))\n",
    "        for fam in fams:\n",
    "            with cd('{}/{}/{}'.format(workdir, gene, fam)):\n",
    "                units = map(lambda x: re.sub(r'^(\\d+?)\\.PRE$', r'\\1', x) ,glob.glob('*.PRE'))\n",
    "                for unit in units:\n",
    "                    copy_file('{}.LOC'.format(unit), 'datafile.dat')\n",
    "                    copy_file('{}.PRE'.format(unit), 'pedfile.pre')\n",
    "\n",
    "                    step1 = runCommand(['makeped', 'pedfile.pre', 'pedfile.ped', 'n'],\n",
    "                                       show_stderr = False, return_zero = False)\n",
    "                    if step1[1]:\n",
    "                        if env.debug:\n",
    "                            with env.lock:\n",
    "                                print(\"makeped error:\", step1[1], file = sys.stderr)\n",
    "                        with env.makeped_counter.get_lock():\n",
    "                            env.makeped_counter.value += 1\n",
    "                        with env.lock:\n",
    "                            errfile.write(step1[1])\n",
    "                        continue\n",
    "                    step2 = runCommand(['pedcheck', '-p', 'pedfile.ped', '-d', 'datafile.dat', '-c'],\n",
    "                                       show_stderr = False, return_zero = False)\n",
    "                    if step2[1]:\n",
    "                        lines = [x for x in step2[1].split('\\n')\n",
    "                                 if not x.strip().startswith('*') and x.strip()]\n",
    "                        if len(lines) > 0:\n",
    "                            env.log('{} lines'.format(len(lines)))\n",
    "                            with env.lock:\n",
    "                                errfile.write(step2[1])\n",
    "                            if env.debug:\n",
    "                                with env.lock:\n",
    "                                    print(\"pedcheck error:\", '\\n'.join(lines), file = sys.stderr)\n",
    "                    try:\n",
    "                        copy_file('zeroout.dat', 'pedfile.dat')\n",
    "                    except:\n",
    "                        with env.pedcheck_counter.get_lock():\n",
    "                            env.pedcheck_counter.value += 1\n",
    "                        continue\n",
    "                    step3 = runCommand('unknown', show_stderr = False, return_zero = False)\n",
    "                    if step3[1]:\n",
    "                        if env.debug:\n",
    "                            with env.lock:\n",
    "                                print(\"unkn error:\", step3[1], file = sys.stderr)\n",
    "                        with env.unknown_counter.get_lock():\n",
    "                            env.unknown_counter.value += 1\n",
    "                        with env.lock:\n",
    "                            errfile.write(step3[1])\n",
    "                        continue\n",
    "                    step4 = runCommand('mlink', show_stderr = False, return_zero = False)\n",
    "                    if step4[1]:\n",
    "                        if env.debug:\n",
    "                            with env.lock:\n",
    "                                print(\"mlink error:\", step4[1], file = sys.stderr)\n",
    "                        with env.lock:\n",
    "                            errfile.write(step4[1])\n",
    "                    try:\n",
    "                        copy_file('outfile.dat', '{}.out'.format(unit))\n",
    "                    except:\n",
    "                        with env.mlink_counter.get_lock():\n",
    "                            env.mlink_counter.value += 1\n",
    "                        continue\n",
    "                    #clean linkage tmp files\n",
    "                    for f in set(glob.glob('*.dat') + glob.glob('ped*') + ['names.tmp']):\n",
    "                        os.remove(f)\n",
    "                    #collect lod scores of different thelta for the fam\n",
    "                    with open('{}.out'.format(unit)) as out:\n",
    "                        raw = out.read()\n",
    "                        for i in re.finditer(r'^THETAS\\s+(0\\.\\d+)(?:\\n.+?){7}LOD SCORE =\\s+(-?\\d+\\.\\d+)', raw, re.MULTILINE):\n",
    "                            theta, lod = map(float, i.group(1,2))\n",
    "                            #if float(lod) < 1e-6:\n",
    "                            #    lod = 0\n",
    "                            if theta not in lods:\n",
    "                                lods[theta] = {fam: lod}\n",
    "                            elif fam not in lods[theta] or lod > lods[theta][fam]:\n",
    "                                lods[theta][fam] = lod\n",
    "        for theta in sorted(lods.keys()):\n",
    "            lods_fh.write('{} {} {} {}\\n'.format(gene, ' '.join(map(str, genemap[gene])), theta, sum(lods[theta].values())))\n",
    "            for fam in lods[theta]:\n",
    "                famlods_fh.write('{} {} {} {} {}\\n'.format(fam, gene, ' '.join(map(str, genemap[gene])), theta, lods[theta][fam]))\n",
    "            print(lods[theta].values())\n",
    "            res = minimize_scalar(hlod_fun(lods[theta].values(), -1), bounds=(0,1), method='bounded', options={'xatol':1e-8})\n",
    "            a = res.x\n",
    "            hlods_fh.write('{} {} {} {} {}\\n'.format(gene, ' '.join(map(str, genemap[gene])), a, theta, hlod_fun(lods[theta].values())(a)))\n",
    "        with env.run_counter.get_lock():\n",
    "            env.run_counter.value += 1\n",
    "        if env.run_counter.value % (env.batch * env.jobs) == 0:\n",
    "            env.log('Linkage analysis for {:,d} units completed {{{:.2%}}} ...'.format(env.run_counter.value, float(env.run_counter.value)/env.success_counter.value), flush=True)\n",
    "    lods_fh.close()\n",
    "    hlods_fh.close()\n",
    "    famlods_fh.close()\n",
    "    if to_plot:\n",
    "        heatmap('{}/heatmap/{}.lods'.format(env.outdir, basename(workdir)), theta_inc, theta_max)\n",
    "        heatmap('{}/heatmap/{}.hlods'.format(env.outdir, basename(workdir)), theta_inc, theta_max)\n",
    "    #env.log(\"Finished running LINKAGE for {}.\".format(workdir), flush=True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_linkage(args.blueprint, args.theta_inc, args.theta_max, args.output_limit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os, subprocess, shutil, glob, shlex, re, hashlib, tempfile\n",
    "try:\n",
    "    from cStringIO import StringIO ## for Python 2\n",
    "    import urlparse\n",
    "except ImportError:\n",
    "    from io import StringIO ## for Python 3\n",
    "    from urllib import parse as urlparse\n",
    "from contextlib import contextmanager\n",
    "from multiprocessing import Pool, Process, Queue, Lock, Value, cpu_count\n",
    "import itertools\n",
    "from collections import OrderedDict, defaultdict, Counter\n",
    "from shutil import rmtree as remove_tree\n",
    "from zipfile import ZipFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpu_count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parmap(f, X, nprocs = cpu_count()):\n",
    "    def spawn(f):\n",
    "        def fun(q_in,q_out):\n",
    "            while True:\n",
    "                i,x = q_in.get()\n",
    "                if i is None:\n",
    "                    break\n",
    "                q_out.put((i,f(x)))\n",
    "        return fun\n",
    "    #\n",
    "    q_in   = Queue(1)\n",
    "    q_out  = Queue()\n",
    "    proc = [Process(target=spawn(f),args=(q_in,q_out)) for _ in range(nprocs)]\n",
    "    for p in proc:\n",
    "        p.daemon = True\n",
    "        p.start()\n",
    "    sent = [q_in.put((i,x)) for i,x in enumerate(X)]\n",
    "    [q_in.put((None,None)) for _ in range(nprocs)]\n",
    "    res = [q_out.get() for _ in range(len(sent))]\n",
    "    [p.join() for p in proc]\n",
    "    return [x for i,x in sorted(res)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1;40;32mMESSAGE: Running linkage analysis ...\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Linkage analysis succesfully performed for 1 units\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: 2 \"pedcheck\" runtime errors occurred\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Report for [tsq20211129] is generated in HTML format\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "if args.run_linkage:\n",
    "    cache.setID('analysis')\n",
    "    if not args.vanilla and cache.check():\n",
    "        env.log('Loading linkage analysis result from archive ...'.format(fmt.upper()))\n",
    "        cache.load(target_dir = env.output, names = ['heatmap'])\n",
    "    else:\n",
    "        env.log('Running linkage analysis ...'.format(fmt.upper()))\n",
    "        run_linkage(args.blueprint, args.theta_inc, args.theta_max, args.output_limit)\n",
    "        env.log('Linkage analysis succesfully performed for {:,d} units\\n'.\\\n",
    "                format(env.run_counter.value, fmt.upper()), flush = True)\n",
    "        if env.makeped_counter.value:\n",
    "            env.log('{} \"makeped\" runtime errors occurred'.format(env.makeped_counter.value))\n",
    "        if env.pedcheck_counter.value:\n",
    "            env.log('{} \"pedcheck\" runtime errors occurred'.format(env.pedcheck_counter.value))\n",
    "        if env.unknown_counter.value:\n",
    "            env.log('{} \"unknown\" runtime errors occurred'.format(env.unknown_counter.value))\n",
    "        if env.mlink_counter.value:\n",
    "            env.log('{} \"mlink\" runtime errors occurred'.format(env.mlink_counter.value))\n",
    "        cache.write(arcroot = 'heatmap', source_dir = os.path.join(env.output, 'heatmap'), mode = 'a')\n",
    "    html(args.theta_inc, args.theta_max, args.output_limit)\n",
    "else:\n",
    "    env.log('Saving data to [{}]'.format(os.path.abspath(env.output)))\n",
    "    cache.load(target_dir = env.output, names = [fmt.upper() for fmt in args.format])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1;40;32mMESSAGE: Saving data to [/mnt/mfs/statgen/yin/Github/linkage/SEQpy2/testseqlink]\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "    env.log('Saving data to [{}]'.format(os.path.abspath(env.output)))\n",
    "    cache.load(target_dir = env.output, names = [fmt.upper() for fmt in args.format])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.Testing heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hinton(filename, max_weight=None, ax=None):\n",
    "    if ax is None:\n",
    "        ax = plt.gca()\n",
    "    matrix = np.random.rand(20, 20) - 0.5\n",
    "    if not max_weight:\n",
    "        max_weight = 2**np.ceil(np.log(np.abs(matrix).max())/np.log(2))\n",
    "    ax.patch.set_facecolor('gray')\n",
    "    ax.set_aspect('equal', 'box')\n",
    "    ax.xaxis.set_major_locator(plt.NullLocator())\n",
    "    ax.yaxis.set_major_locator(plt.NullLocator())\n",
    "    chrID = re.search(r'\\.chr([0-9XY]+)\\.', filename).group(1)\n",
    "    ax.set_title('Chromosome {}'.format(chrID))\n",
    "    for (x,y),w in np.ndenumerate(matrix):\n",
    "        color = 'white' if w > 0 else 'black'\n",
    "        size = np.sqrt(np.abs(w))\n",
    "        rect = plt.Rectangle([x - size / 2, y - size / 2], size, size,\n",
    "                             facecolor=color, edgecolor=color)\n",
    "        ax.add_patch(rect)\n",
    "    ax.autoscale_view()\n",
    "    ax.invert_yaxis()\n",
    "    plt.savefig(filename)\n",
    "\n",
    "\n",
    "def heatmap(file, theta_inc, theta_max):\n",
    "    #env.log(\"Start ploting heatmap for {} ...\".format(file), flush=True)\n",
    "    if os.path.getsize(file) == 0:\n",
    "        hinton('{}.png'.format(file))\n",
    "        return\n",
    "    lods = []\n",
    "    with open(file, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            theta,lod = line.split()[-2:]\n",
    "            if float(theta) >= theta_max:\n",
    "                print(line)\n",
    "                continue\n",
    "            lods.append(lod)\n",
    "        #add random number.\n",
    "        if max(lods) == min(lods):\n",
    "            #env.log('Max equals Min for [{}], No real heatmap will be generated.'.format(file))\n",
    "            hinton('{}.png'.format(file))\n",
    "            return\n",
    "        Num=int(round(theta_max/theta_inc))\n",
    "        lods = np.array(list(map(float,lods))).reshape((-1,Num))\n",
    "        chrID = re.search(r'\\.chr([0-9XY]+)\\.', file).group(1)\n",
    "        fig, ax = plt.subplots(1)\n",
    "        ax.set_title('Chromosome {}'.format(chrID))\n",
    "        plt.pcolormesh(fig,ax,lods.transpose(),\n",
    "                       xticklabels=[''] * len(lods),\n",
    "                       yticklabels=np.round(np.array(range(Num)) * theta_inc,2).tolist(),\n",
    "                       cmap=brewer2mpl.get_map('Blues', 'Sequential', 9).mpl_colormap)\n",
    "        fig.savefig('{}.png'.format(file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.05"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "theta_inc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Num=int(round(theta_max/theta_inc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lods = ['0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lods = np.array(list(map(float, lods))).reshape((-1,Num))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4, 10)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lods.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(lods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['', '', '', '']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[''] * len(lods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.0, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.round(np.array(range(Num)) * theta_inc,2).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'LinearSegmentedColormap' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/2206534.1.plot.q/ipykernel_30528/241223039.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mbrewer2mpl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Blues'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Sequential'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m9\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmpl_colormap\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: 'LinearSegmentedColormap' object is not subscriptable"
     ]
    }
   ],
   "source": [
    "brewer2mpl.get_map('Blues', 'Sequential', 9).mpl_colormap[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lods = np.random.rand(4,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.QuadMesh at 0x2b7a0b1f2be0>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAARAElEQVR4nO3df6zddX3H8efLtpQf8ktggm2xJDAcMhVpEEeyEJgJKKFm1g0zERiki5EJi4kRk8EkWTIzo85hJA0wURhiKnOV4A8iGDUL1VL5XZz1Z9vVUQqCID8svvfH/Wpv7m455957br/1c5+P5KTfH5/zPa9+0/u6337P93tOqgpJUrte0ncASdLssuglqXEWvSQ1zqKXpMZZ9JLUOItekho3dNEnmZfke0lunWTdwiQ3J9mYZG2SpSNNKUmatqkc0V8CbNjFuguBx6vqaOBjwIdnGkySNBpDFX2SxcBbgGt2MWQ5cH03vRo4PUlmHk+SNFPzhxz3ceD9wP67WL8I2ARQVTuSPAEcAjw6flCSlcBKgPn7zD/x4KUHTCPyaD3zwoK+I7Djif4zANSw/xpm0V5P7iF3ah/x674TkE3z+o4AwLOH9Z9j4c+e7jsCAHv/Uf/Hr9s2PPZoVR02lecM/NFOchbwSFXdneTUaWYDoKpWAasA/uC4Q2rFDWfOZHMj8dDjh/cdgW1fW9R3BACeP6jvBHDk7c/2HWHM32/rOwELLt2v7wgAPPw3B/YdgWMuXtt3BABedcNefUfgqhP//adTfc4wp25OAc5O8hPgc8BpSW6YMGYLsAQgyXzgQGD7VMNIkkZvYNFX1WVVtbiqlgLnAHdU1TsnDFsDnNdNr+jG7CH/B5ekuW3aZ2WTXAmsq6o1wLXAZ5NsBB5j7BeCJGkPMKWir6pvAN/opi8ft/xZ4O2jDCZJGg3vjJWkxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGWfSS1LiBRZ9k7yTfSXJvkgeTfGiSMecn2Zbknu5x0ezElSRN1TDfGfsccFpVPZVkAfDtJF+uqrsmjLu5qi4efURJ0kwMLPqqKuCpbnZB96jZDCVJGp2hztEnmZfkHuAR4PaqWjvJsLcluS/J6iRLRhlSkjR9QxV9Vb1QVa8DFgMnJTl+wpAvAUur6jXA7cD1k20nycok65Kse+bxZ2cQW5I0rClddVNVvwDuBM6YsHx7VT3XzV4DnLiL56+qqmVVtWyfg/eeRlxJ0lQNc9XNYUkO6qb3Ad4EPDxhzBHjZs8GNowwoyRpBoa56uYI4Pok8xj7xfD5qro1yZXAuqpaA7w3ydnADuAx4PzZCixJmpphrrq5DzhhkuWXj5u+DLhstNEkSaPgnbGS1DiLXpIaZ9FLUuMseklqnEUvSY2z6CWpcRa9JDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4wYWfZK9k3wnyb1JHkzyoUnGLExyc5KNSdYmWToraSVJUzbMEf1zwGlV9VrgdcAZSU6eMOZC4PGqOhr4GPDhkaaUJE3bwKKvMU91swu6R00Ythy4vpteDZyeJCNLKUmatvnDDEoyD7gbOBr4ZFWtnTBkEbAJoKp2JHkCOAR4dMJ2VgIrAY5cNJ+rXvGdmaUfgZMfWdF3BA5f+0zfEQD44XlD/XOYVT9+6159RwBg/vaD+45Arph4PNWPPzxnfd8R4KQ/7jsBAP+98jd9R5iWod6MraoXqup1wGLgpCTHT+fFqmpVVS2rqmWHHTJvOpuQJE3RlK66qapfAHcCZ0xYtQVYApBkPnAgsH0E+SRJMzTMVTeHJTmom94HeBPw8IRha4DzuukVwB1VtWf8v1OS5rhhTsoeAVzfnad/CfD5qro1yZXAuqpaA1wLfDbJRuAx4JxZSyxJmpKBRV9V9wEnTLL88nHTzwJvH200SdIoeGesJDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGDfPl4EuS3JnkoSQPJrlkkjGnJnkiyT3d4/LJtiVJ2v2G+XLwHcD7qmp9kv2Bu5PcXlUPTRj3rao6a/QRJUkzMfCIvqq2VtX6bvqXwAZg0WwHkySNxpTO0SdZCpwArJ1k9RuT3Jvky0levYvnr0yyLsm6bdtfmHpaSdKUDV30SV4KfAG4tKqenLB6PfDKqnot8K/AFyfbRlWtqqplVbXssEPmTTOyJGkqhir6JAsYK/kbq+qWieur6smqeqqbvg1YkOTQkSaVJE3LMFfdBLgW2FBVH93FmMO7cSQ5qdvu9lEGlSRNzzBX3ZwCnAvcn+SebtkHgSMBqupqYAXw7iQ7gGeAc6qqRh9XkjRVA4u+qr4NZMCYq4CrRhVKkjQ63hkrSY2z6CWpcRa9JDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxw3w5+JIkdyZ5KMmDSS6ZZEySfCLJxiT3JXn97MSVJE3VMF8OvgN4X1WtT7I/cHeS26vqoXFjzgSO6R5vAD7V/SlJ6tnAI/qq2lpV67vpXwIbgEUThi0HPlNj7gIOSnLEyNNKkqZsmCP630myFDgBWDth1SJg07j5zd2yrROevxJYCbBw4UH82V/99RTjjt4b/vn+viPw/ede1XcEAPL0wr4j8MK+v+k7AgDHvntL3xHY8JGj+o4AwD/84K6+I/ChVy/oOwIAF937QN8R+NrRU3/O0G/GJnkp8AXg0qp6cuovBVW1qqqWVdWyvfbabzqbkCRN0VBFn2QBYyV/Y1XdMsmQLcCScfOLu2WSpJ4Nc9VNgGuBDVX10V0MWwO8q7v65mTgiarauouxkqTdaJhz9KcA5wL3J7mnW/ZB4EiAqroauA14M7AR+BVwwciTSpKmZWDRV9W3gQwYU8B7RhVKkjQ63hkrSY2z6CWpcRa9JDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxw3w5+HVJHknywC7Wn5rkiST3dI/LRx9TkjRdw3w5+KeBq4DPvMiYb1XVWSNJJEkaqYFH9FX1TeCx3ZBFkjQLRnWO/o1J7k3y5SSv3tWgJCuTrEuy7vnnnx7RS0uSXswwp24GWQ+8sqqeSvJm4IvAMZMNrKpVwCqAAw5YXCN4bUnSADM+oq+qJ6vqqW76NmBBkkNnnEySNBIzLvokhydJN31St83tM92uJGk0Bp66SXITcCpwaJLNwBXAAoCquhpYAbw7yQ7gGeCcqvK0jCTtIQYWfVW9Y8D6qxi7/FKStAfyzlhJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGWfSS1DiLXpIaZ9FLUuMseklqnEUvSY0bWPRJrkvySJIHdrE+ST6RZGOS+5K8fvQxJUnTNcwR/aeBM15k/ZnAMd1jJfCpmceSJI3KwKKvqm8Cj73IkOXAZ2rMXcBBSY4YVUBJ0szMH8E2FgGbxs1v7pZtnTgwyUrGjvpZcNiBbLvkmRG8/MwsnLej7wi85AebBg/aDe7986/0HYELfnJW3xEA2LDy2L4j8LL/6jvBmEtffk7fEfjH+/+j7wgAXHLvX/YdAVg/5Wfs1jdjq2pVVS2rqmXzDth3d760JM1Zoyj6LcCScfOLu2WSpD3AKIp+DfCu7uqbk4Enqur/nbaRJPVj4Dn6JDcBpwKHJtkMXAEsAKiqq4HbgDcDG4FfARfMVlhJ0tQNLPqqeseA9QW8Z2SJJEkj5Z2xktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUOItekhpn0UtS4yx6SWqcRS9JjbPoJalxFr0kNc6il6TGWfSS1DiLXpIaN1TRJzkjyfeTbEzygUnWn59kW5J7usdFo48qSZqOYb4cfB7wSeBNwGbgu0nWVNVDE4beXFUXz0JGSdIMDHNEfxKwsap+VFXPA58Dls9uLEnSqAxT9IuATePmN3fLJnpbkvuSrE6yZCTpJEkzNqo3Y78ELK2q1wC3A9dPNijJyiTrkqx74clfjeilJUkvZpii3wKMP0Jf3C37naraXlXPdbPXACdOtqGqWlVVy6pq2bwD9p1OXknSFA1T9N8FjklyVJK9gHOANeMHJDli3OzZwIbRRZQkzcTAq26qakeSi4GvAvOA66rqwSRXAuuqag3w3iRnAzuAx4DzZzGzJGkKBhY9QFXdBtw2Ydnl46YvAy4bbTRJ0ih4Z6wkNc6il6TGWfSS1DiLXpIaZ9FLUuMseklqnEUvSY2z6CWpcRa9JDXOopekxln0ktQ4i16SGmfRS1LjLHpJapxFL0mNs+glqXEWvSQ1zqKXpMZZ9JLUuKGKPskZSb6fZGOSD0yyfmGSm7v1a5MsHXlSSdK0DCz6JPOATwJnAscB70hy3IRhFwKPV9XRwMeAD486qCRpeoY5oj8J2FhVP6qq54HPAcsnjFkOXN9NrwZOT5LRxZQkTVeq6sUHJCuAM6rqom7+XOANVXXxuDEPdGM2d/M/7MY8OmFbK4GV3ezxwAOj+ov8njsUeHTgqLnBfbGT+2In98VOx1bV/lN5wvzZSjKZqloFrAJIsq6qlu3O199TuS92cl/s5L7YyX2xU5J1U33OMKdutgBLxs0v7pZNOibJfOBAYPtUw0iSRm+Yov8ucEySo5LsBZwDrJkwZg1wXje9ArijBp0TkiTtFgNP3VTVjiQXA18F5gHXVdWDSa4E1lXVGuBa4LNJNgKPMfbLYJBVM8jdGvfFTu6LndwXO7kvdpryvhj4Zqwk6febd8ZKUuMseklqXC9FP+gjFeaKJEuS3JnkoSQPJrmk70x9SjIvyfeS3Np3lr4lOSjJ6iQPJ9mQ5I19Z+pLkr/rfj4eSHJTkr37zrS7JLkuySPdvUq/XfayJLcn+UH358GDtrPbi37Ij1SYK3YA76uq44CTgffM4X0BcAmwoe8Qe4h/Ab5SVa8CXssc3S9JFgHvBZZV1fGMXRAyzMUerfg0cMaEZR8Avl5VxwBf7+ZfVB9H9MN8pMKcUFVbq2p9N/1Lxn6YF/Wbqh9JFgNvAa7pO0vfkhwI/CljV7NRVc9X1S96DdWv+cA+3T06+wL/03Oe3aaqvsnYlYzjjf/ImeuBtw7aTh9FvwjYNG5+M3O03MbrPvHzBGBtz1H68nHg/cBves6xJzgK2Ab8W3cq65ok+/Udqg9VtQX4CPAzYCvwRFV9rd9UvXt5VW3tpn8OvHzQE3wzdg+Q5KXAF4BLq+rJvvPsbknOAh6pqrv7zrKHmA+8HvhUVZ0APM0Q/z1vUXf+eTljv/xeAeyX5J39ptpzdDemDrxGvo+iH+YjFeaMJAsYK/kbq+qWvvP05BTg7CQ/YexU3mlJbug3Uq82A5ur6rf/u1vNWPHPRX8G/LiqtlXVr4FbgD/pOVPf/jfJEQDdn48MekIfRT/MRyrMCd1HOV8LbKiqj/adpy9VdVlVLa6qpYz9e7ijqubsUVtV/RzYlOTYbtHpwEM9RurTz4CTk+zb/byczhx9Y3qc8R85cx7wn4OesFs/vRJ2/ZEKuzvHHuIU4Fzg/iT3dMs+WFW39RdJe4i/BW7sDoZ+BFzQc55eVNXaJKuB9YxdpfY95tDHISS5CTgVODTJZuAK4J+Azye5EPgp8BcDt+NHIEhS23wzVpIaZ9FLUuMseklqnEUvSY2z6CWpcRa9JDXOopekxv0fueZpCsRwN5QAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.pcolormesh(lods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make these smaller to increase the resolution\n",
    "dx, dy = 1, 0.05\n",
    "\n",
    "# generate 2 2d grids for the x & y bounds\n",
    "y, x = np.mgrid[slice(0, 0.5 + dy, dy),\n",
    "                slice(0, 4 + dx, dx)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.QuadMesh at 0x2b7a0b397370>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXsAAAD8CAYAAACW/ATfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAAPP0lEQVR4nO3dbYxcZ32G8ev2JiZAQlJqmqS289LWTWUgCcE4qUAQpSA5UNmopKpT0RIEMi11oQQVkrZKVfcTVKIIal6sECl9ARNS1C7UKEUlFFEpqQOEFyc1XSKKnUYNBEhogRjDvx92AsNqlz27O+uzO8/1k1aac84zc+597HPv2TM7M6kqJEnjbU3fASRJy8+yl6QGWPaS1ADLXpIaYNlLUgMse0lqQKeyT7ItyeEkU0mum2X7NUm+muTuwdcrRx9VkrRYJ803IMkEsBd4AXAUOJhksqrumTH0/VW1exkySpKWqMuZ/VZgqqruq6pjwH5gx/LGkiSN0rxn9sB64MjQ8lHg0lnGvSTJc4EvAq+rqiMzByTZBewCWLP28c98wpnnLDyxZvV/3/pu3xHGxlPPeXLfEcbK4Qce6TvCWPneV7/0tap6ykLv16Xsu/gQ8L6qejTJq4CbgStmDqqqfcA+gNPO+aXa8vqbRrR73fGv9/YdYWx88J2/2XeEsXLFnn/uO8JYOfqOF//XYu7X5TLO/cDGoeUNg3U/VFUPVdWjg8UbgWcuJowkaXl0KfuDwKYk5ydZC+wEJocHJDl7aHE74GmmJK0g817GqarjSXYDtwETwE1VdSjJHuCuqpoEXpNkO3Ac+DpwzTJmliQtUKdr9lV1ADgwY90NQ7evB64fbTRJ0qj4ClpJaoBlL0kNsOwlqQGWvSQ1wLKXpAZY9pLUAMtekhpg2UtSAyx7SWqAZS9JDbDsJakBlr0kNcCyl6QGWPaS1ADLXpIaYNlLUgNSVb3seM0Z59TjnvOGXvY9jh689dV9Rxgb//3N7/YdYaycu+4JfUcYK48/OZ+qqi0LvZ9n9pLUAMtekhpg2UtSAyx7SWqAZS9JDbDsJakBlr0kNcCyl6QGWPaS1ADLXpIaYNlLUgMse0lqgGUvSQ2w7CWpAZa9JDXAspekBlj2ktQAy16SGtCp7JNsS3I4yVSS637CuJckqSQL/sgsSdLymbfsk0wAe4Ergc3A1Uk2zzLuNOC1wJ2jDilJWpouZ/Zbgamquq+qjgH7gR2zjPtz4E2An9YsSSvMSR3GrAeODC0fBS4dHpDkEmBjVf1Tkj+c64GS7AJ2Aaw940wues7TF55Ys/qZF7+17whj46HJ1/UdYaz8/ge/0HcEMYInaJOsAd4CvH6+sVW1r6q2VNWWk594+lJ3LUnqqEvZ3w9sHFreMFj3mNOApwEfT/Jl4DJg0idpJWnl6FL2B4FNSc5PshbYCUw+trGqHq6qdVV1XlWdB9wBbK+qu5YlsSRpweYt+6o6DuwGbgPuBW6pqkNJ9iTZvtwBJUlL1+UJWqrqAHBgxrob5hh7+dJjSZJGyVfQSlIDLHtJaoBlL0kNsOwlqQGWvSQ1wLKXpAZY9pLUAMtekhpg2UtSAyx7SWqAZS9JDbDsJakBlr0kNcCyl6QGWPaS1ADLXpIa0OnDS5bDWaefwhtfeEFfux87+89+Ut8RxsaOfXf2HWGsvPs3Lu47wli5cZH388xekhpg2UtSAyx7SWqAZS9JDbDsJakBlr0kNcCyl6QGWPaS1ADLXpIaYNlLUgMse0lqgGUvSQ2w7CWpAZa9JDXAspekBlj2ktQAy16SGtCp7JNsS3I4yVSS62bZ/jtJPp/k7iSfTLJ59FElSYs1b9knmQD2AlcCm4GrZynz91bV06vqYuDNwFtGHVSStHhdzuy3AlNVdV9VHQP2AzuGB1TVI0OLTwRqdBElSUvV5QPH1wNHhpaPApfOHJTk94BrgbXAFbM9UJJdwC6AnzrzZznyyHcWmldzeOdVF/YdYWxcffNdfUcYK7+299/6jiBG+ARtVe2tqp8H3gj8yRxj9lXVlqracuoZTx7VriVJ8+hS9vcDG4eWNwzWzWU/8OIlZJIkjViXsj8IbEpyfpK1wE5gcnhAkk1Diy8C/nN0ESVJSzXvNfuqOp5kN3AbMAHcVFWHkuwB7qqqSWB3kucD3wO+AbxsOUNLkhamyxO0VNUB4MCMdTcM3X7tiHNJkkbIV9BKUgMse0lqgGUvSQ2w7CWpAZa9JDXAspekBlj2ktQAy16SGmDZS1IDLHtJaoBlL0kNsOwlqQGWvSQ1wLKXpAZY9pLUAMtekhqQquplx6ec9Yt17sve3su+x9HWC8/qO8LYeNIT1vYdYay8asvG+Qeps4vOOe1TVbVloffzzF6SGmDZS1IDLHtJaoBlL0kNsOwlqQGWvSQ1wLKXpAZY9pLUAMtekhpg2UtSAyx7SWqAZS9JDbDsJakBlr0kNcCyl6QGWPaS1ADLXpIaYNlLUgM6lX2SbUkOJ5lKct0s269Nck+SzyX5lyTnjj6qJGmx5i37JBPAXuBKYDNwdZLNM4Z9BthSVRcCtwJvHnVQSdLidTmz3wpMVdV9VXUM2A/sGB5QVbdX1bcHi3cAG0YbU5K0FCd1GLMeODK0fBS49CeMfwXwkdk2JNkF7AJg7Wl85faPdkupeV3+rJf3HWFsfP1bj/YdYax84NADfUcQ3cq+syQvBbYAz5tte1XtA/YBrHniWTXKfUuS5tal7O8HNg4tbxis+zFJng/8MfC8qvLUSJJWkC7X7A8Cm5Kcn2QtsBOYHB6Q5BnAu4HtVfXg6GNKkpZi3rKvquPAbuA24F7glqo6lGRPku2DYX8BnAp8IMndSSbneDhJUg86XbOvqgPAgRnrbhi6/fwR55IkjZCvoJWkBlj2ktQAy16SGmDZS1IDLHtJaoBlL0kNsOwlqQGWvSQ1wLKXpAZY9pLUAMtekhpg2UtSAyx7SWqAZS9JDbDsJakBlr0kNWCkHzi+EBddsIGPffxNfe1+7PzPw37s76gc//4P+o4wVj555Gt9RxCe2UtSEyx7SWqAZS9JDbDsJakBlr0kNcCyl6QGWPaS1ADLXpIaYNlLUgMse0lqgGUvSQ2w7CWpAZa9JDXAspekBlj2ktQAy16SGmDZS1IDOpV9km1JDieZSnLdLNufm+TTSY4nuWr0MSVJSzFv2SeZAPYCVwKbgauTbJ4x7CvANcB7Rx1QkrR0XT6DdiswVVX3ASTZD+wA7nlsQFV9ebDND++UpBWoS9mvB44MLR8FLl3MzpLsAnYBTJy6jqde+4+LeRjNYmJiou8IY2P9xjP6jjBWjh37ft8RxAl+graq9lXVlqrasuaU00/kriWpaV3K/n5g49DyhsE6SdIq0aXsDwKbkpyfZC2wE5hc3liSpFGat+yr6jiwG7gNuBe4paoOJdmTZDtAkmclOQr8OvDuJIeWM7QkaWG6PEFLVR0ADsxYd8PQ7YNMX96RJK1AvoJWkhpg2UtSAyx7SWqAZS9JDbDsJakBlr0kNcCyl6QGWPaS1ADLXpIaYNlLUgMse0lqgGUvSQ2w7CWpAZa9JDXAspekBlj2ktSATh9eshwed8pJbLrgzL52P3bedtWFfUcYG5dtv77vCGNl3bNf0HcE4Zm9JDXBspekBlj2ktQAy16SGmDZS1IDLHtJaoBlL0kNsOwlqQGWvSQ1wLKXpAZY9pLUAMtekhpg2UtSAyx7SWqAZS9JDbDsJakBlr0kNcCyl6QGdCr7JNuSHE4yleS6WbY/Lsn7B9vvTHLeyJNKkhZt3rJPMgHsBa4ENgNXJ9k8Y9grgG9U1S8Afwm8adRBJUmL1+XMfiswVVX3VdUxYD+wY8aYHcDNg9u3Ar+SJKOLKUlaipM6jFkPHBlaPgpcOteYqjqe5GHgp4GvDQ9KsgvYNVh89BPXPucLiwl9gq1jxvexEl187erIyeqYz9WQEVZJzqN3710VOVkl8wlcsJg7dSn7kamqfcA+gCR3VdWWE7n/xTDnaK2GnKshI5hz1FZTzsXcr8tlnPuBjUPLGwbrZh2T5CTgdOChxQSSJI1el7I/CGxKcn6StcBOYHLGmEngZYPbVwEfq6oaXUxJ0lLMexlncA1+N3AbMAHcVFWHkuwB7qqqSeA9wN8kmQK+zvQPhPnsW0LuE8mco7Uacq6GjGDOURvrnPEEXJLGn6+glaQGWPaS1IBlL/vV8lYLHXJek+SrSe4efL2yh4w3JXkwyayvT8i0tw2+h88lueREZxzkmC/n5UkeHprLG3rIuDHJ7UnuSXIoyWtnGdP7fHbMuRLm85Qk/57ks4OcfzbLmN6P9Y45ez/WBzkmknwmyYdn2bbwuayqZfti+gndLwE/B6wFPgtsnjHm1cC7Brd3Au9fzkxLyHkN8FcnOtuMDM8FLgG+MMf2FwIfAQJcBty5QnNeDny457k8G7hkcPs04Iuz/Jv3Pp8dc66E+Qxw6uD2ycCdwGUzxqyEY71Lzt6P9UGOa4H3zvZvu5i5XO4z+9XyVgtdcvauqj7B9F87zWUH8Nc17Q7gjCRnn5h0P9IhZ++q6oGq+vTg9reAe5l+Jfiw3uezY87eDebofweLJw++Zv71R+/HesecvUuyAXgRcOMcQxY8l8td9rO91cLM/6g/9lYLwGNvtXAidckJ8JLBr/O3Jtk4y/a+df0+VoJfHvwq/ZEkT+0zyOBX4GcwfZY3bEXN50/ICStgPgeXHe4GHgQ+WlVzzmePx3qXnND/sf5W4A3AD+bYvuC59Ana7j4EnFdVFwIf5Uc/VbVwnwbOraqLgLcD/9BXkCSnAn8P/EFVPdJXjvnMk3NFzGdVfb+qLmb6VfZbkzytjxzz6ZCz12M9ya8CD1bVp0b5uMtd9qvlrRbmzVlVD1XVo4PFG4FnnqBsC9FlvntXVY889qt0VR0ATk6y7kTnSHIy0wX6d1X1wVmGrIj5nC/nSpnPoTzfBG4Hts3YtBKO9R+aK+cKONafDWxP8mWmLylfkeRvZ4xZ8Fwud9mvlrdamDfnjGu125m+drrSTAK/PfgrksuAh6vqgb5DzZTkrMeuLybZyvT/wxN60A/2/x7g3qp6yxzDep/PLjlXyHw+JckZg9uPB14A/MeMYb0f611y9n2sV9X1VbWhqs5juos+VlUvnTFswXO5rO96Wcv3Vgt95HxNku3A8UHOa050ziTvY/ovL9YlOQr8KdNPMFFV7wIOMP0XJFPAt4GXn+iMHXNeBfxukuPAd4CdPfyAfzbwW8DnB9dvAf4IOGco50qYzy45V8J8ng3cnOkPO1oD3FJVH15px3rHnL0f67NZ6lz6dgmS1ACfoJWkBlj2ktQAy16SGmDZS1IDLHtJaoBlL0kNsOwlqQH/D4Y+KAQVt7iYAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.pcolormesh(x,y,lods.transpose(),cmap=brewer2mpl.get_map('Blues', 'Sequential', 9).mpl_colormap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "tuple index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/2206534.1.plot.q/ipykernel_30528/3434994219.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_title\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Chromosome {}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrID\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m plt.pcolormesh(fig,ax,lods.transpose(),\n\u001b[0m\u001b[1;32m      6\u001b[0m                \u001b[0mxticklabels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m''\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlods\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m                \u001b[0myticklabels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mround\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mNum\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mtheta_inc\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/seqpy3v0/lib/python3.9/site-packages/matplotlib/pyplot.py\u001b[0m in \u001b[0;36mpcolormesh\u001b[0;34m(alpha, norm, cmap, vmin, vmax, shading, antialiased, data, *args, **kwargs)\u001b[0m\n\u001b[1;32m   2716\u001b[0m         \u001b[0mvmax\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshading\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mantialiased\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2717\u001b[0m         **kwargs):\n\u001b[0;32m-> 2718\u001b[0;31m     __ret = gca().pcolormesh(\n\u001b[0m\u001b[1;32m   2719\u001b[0m         \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malpha\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0malpha\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnorm\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnorm\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcmap\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcmap\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvmin\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvmin\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2720\u001b[0m         \u001b[0mvmax\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvmax\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshading\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mshading\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mantialiased\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mantialiased\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/seqpy3v0/lib/python3.9/site-packages/matplotlib/__init__.py\u001b[0m in \u001b[0;36minner\u001b[0;34m(ax, data, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1410\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0minner\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1411\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1412\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msanitize_sequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1413\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1414\u001b[0m         \u001b[0mbound\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnew_sig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/seqpy3v0/lib/python3.9/site-packages/matplotlib/axes/_axes.py\u001b[0m in \u001b[0;36mpcolormesh\u001b[0;34m(self, alpha, norm, cmap, vmin, vmax, shading, antialiased, *args, **kwargs)\u001b[0m\n\u001b[1;32m   6014\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetdefault\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'edgecolors'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'none'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6015\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6016\u001b[0;31m         X, Y, C, shading = self._pcolorargs('pcolormesh', *args,\n\u001b[0m\u001b[1;32m   6017\u001b[0m                                             shading=shading, kwargs=kwargs)\n\u001b[1;32m   6018\u001b[0m         \u001b[0mcoords\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mY\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/seqpy3v0/lib/python3.9/site-packages/matplotlib/axes/_axes.py\u001b[0m in \u001b[0;36m_pcolorargs\u001b[0;34m(self, funcname, shading, *args, **kwargs)\u001b[0m\n\u001b[1;32m   5507\u001b[0m                             f'but {len(args)} were given')\n\u001b[1;32m   5508\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5509\u001b[0;31m         \u001b[0mNx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mX\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5510\u001b[0m         \u001b[0mNy\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mY\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5511\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mX\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mX\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: tuple index out of range"
     ]
    }
   ],
   "source": [
    "file = '{}/heatmap/{}.lods'.format(env.outdir, basename(workdirs[0]))\n",
    "chrID = re.search(r'\\.chr([0-9XY]+)\\.', file).group(1)\n",
    "fig, ax = plt.subplots(1)\n",
    "ax.set_title('Chromosome {}'.format(chrID))\n",
    "plt.pcolormesh(fig,ax,lods.transpose(),\n",
    "               xticklabels=[''] * len(lods),\n",
    "               yticklabels=np.round(np.array(range(Num)) * theta_inc,2).tolist(),\n",
    "               cmap=brewer2mpl.get_map('Blues', 'Sequential', 9).mpl_colormap)\n",
    "fig.savefig('{}.png'.format(file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOC102725121@1 1 11868 14362 0.5 0.0\n",
      "\n",
      "DDX11L1 1 11873 14409 0.5 0.0\n",
      "\n",
      "WASH7P 1 14361 29370 0.5 0.0\n",
      "\n",
      "MIR6859-1@1,MIR6859-2@1,MIR6859-3@1,MIR6859-4@1 1 17368 17436 0.5 0.0\n",
      "\n",
      "['0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0']\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "cannot reshape array of size 1 into shape (10)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/2206534.1.plot.q/ipykernel_30528/3244704223.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mheatmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'{}/heatmap/{}.lods'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0menv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutdir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbasename\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mworkdirs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtheta_inc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtheta_max\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;31m#heatmap('{}/heatmap/{}.hlods'.format(env.outdir, basename(workdir)), theta_inc, theta_max)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/2206534.1.plot.q/ipykernel_30528/1950724175.py\u001b[0m in \u001b[0;36mheatmap\u001b[0;34m(file, theta_inc, theta_max)\u001b[0m\n\u001b[1;32m     41\u001b[0m             \u001b[0;31m#return\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m         \u001b[0mNum\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mround\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtheta_max\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0mtheta_inc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 43\u001b[0;31m         \u001b[0mlods\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfloat\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlods\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mNum\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     44\u001b[0m         \u001b[0mchrID\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mre\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msearch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mr'\\.chr([0-9XY]+)\\.'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroup\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m         \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: cannot reshape array of size 1 into shape (10)"
     ]
    }
   ],
   "source": [
    "heatmap('{}/heatmap/{}.lods'.format(env.outdir, basename(workdirs[0])), theta_inc, theta_max)\n",
    "#heatmap('{}/heatmap/{}.hlods'.format(env.outdir, basename(workdir)), theta_inc, theta_max)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['./LINKAGE/LINKAGE/LINKAGE.chr1']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "workdirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
