---

title: Linkage module

keywords: fastai
sidebar: home_sidebar

summary: "The core functions for linkage analysis. Paramlink2 is used to do linkage analysis. The R code is bridged to python through rpy2. It run linkage analysis from batch to batch (default is 25 genes per batch). "
description: "The core functions for linkage analysis. Paramlink2 is used to do linkage analysis. The R code is bridged to python through rpy2. It run linkage analysis from batch to batch (default is 25 genes per batch). "
nb_path: "nbs/02_Linkage.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/02_Linkage.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Functions-for-preprocessing-genotypes-or-phased-haplotypes">Functions for preprocessing genotypes or phased haplotypes<a class="anchor-link" href="#Functions-for-preprocessing-genotypes-or-phased-haplotypes"> </a></h3><ul>
<li>input haps</li>
<li>output preprocessed inputs of chp markers</li>
</ul>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="update_haps_ped" class="doc_header"><code>update_haps_ped</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L35" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>update_haps_ped</code>(<strong><code>genes</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="hap2chp" class="doc_header"><code>hap2chp</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L47" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>hap2chp</code>(<strong><code>haps</code></strong>, <strong><code>gene</code></strong>)</p>
</blockquote>
<p>input hap is [varnames,freqs,halpotypes]
output is [chp_varnames,chp_freqs,chps]</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_marker" class="doc_header"><code>generate_marker</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L67" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_marker</code>(<strong><code>alleles</code></strong>)</p>
</blockquote>
<p>array of 0,1,2. 0 if all 0 -&gt; 2 if any 2 else 1</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="recombination_pos" class="doc_header"><code>recombination_pos</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L76" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>recombination_pos</code>(<strong><code>hap</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="recombination_region" class="doc_header"><code>recombination_region</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L83" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>recombination_region</code>(<strong><code>hap</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h4 id="test">test<a class="anchor-link" href="#test"> </a></h4>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;../data/wg20220311/chr19test/CACHE/chr19test43.pickle&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">haps</span><span class="o">=</span><span class="n">genes</span><span class="p">[</span><span class="s1">&#39;APOE&#39;</span><span class="p">][</span><span class="s1">&#39;predata&#39;</span><span class="p">][</span><span class="s1">&#39;1007&#39;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">hap</span><span class="o">=</span><span class="n">haps</span><span class="p">[</span><span class="mi">2</span><span class="p">][:,</span><span class="mi">2</span><span class="p">:]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">hap</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>array([[&#39;2:&#39;, &#39;1:&#39;, &#39;A1,2:&#39;, &#39;1:&#39;],
       [&#39;2:&#39;, &#39;2:&#39;, &#39;1:&#39;, &#39;2:&#39;],
       [&#39;1:&#39;, &#39;1:&#39;, &#39;A2,1:&#39;, &#39;1:&#39;],
       [&#39;2:&#39;, &#39;2:&#39;, &#39;1:&#39;, &#39;2:&#39;],
       [&#39;2:&#39;, &#39;1:&#39;, &#39;A1,2:&#39;, &#39;1:&#39;],
       [&#39;?:&#39;, &#39;?:&#39;, &#39;?:&#39;, &#39;?:&#39;],
       [&#39;2:&#39;, &#39;1/&#39;, &#39;A2,1|&#39;, &#39;1\\&#39;],
       [&#39;2:&#39;, &#39;1/&#39;, &#39;A1,2|&#39;, &#39;2\\&#39;],
       [&#39;2:&#39;, &#39;2|&#39;, &#39;1|&#39;, &#39;2|&#39;],
       [&#39;2:&#39;, &#39;2|&#39;, &#39;1|&#39;, &#39;2|&#39;],
       [&#39;1:&#39;, &#39;1|&#39;, &#39;A2,1|&#39;, &#39;1|&#39;],
       [&#39;2:&#39;, &#39;1|&#39;, &#39;A1,2|&#39;, &#39;1|&#39;],
       [&#39;1:&#39;, &#39;1|&#39;, &#39;A2,1|&#39;, &#39;1|&#39;],
       [&#39;2:&#39;, &#39;1|&#39;, &#39;A1,2|&#39;, &#39;1|&#39;],
       [&#39;1:&#39;, &#39;1|&#39;, &#39;A2,1|&#39;, &#39;1|&#39;],
       [&#39;2:&#39;, &#39;1|&#39;, &#39;A1,2|&#39;, &#39;1|&#39;],
       [&#39;1:&#39;, &#39;1|&#39;, &#39;A2,1|&#39;, &#39;1|&#39;],
       [&#39;2:&#39;, &#39;1|&#39;, &#39;A1,2|&#39;, &#39;1|&#39;]], dtype=&#39;&lt;U7&#39;)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">recombination_pos</span><span class="p">(</span><span class="n">hap</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[0, 1, 3]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">recombination_region</span><span class="p">(</span><span class="n">hap</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[(0, 1), (1, 3)]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">hap2chp</span><span class="p">(</span><span class="n">haps</span><span class="p">,</span><span class="s1">&#39;APOE&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[array([&#39;APOE:0&#39;, &#39;APOE:1&#39;], dtype=&#39;&lt;U6&#39;),
 array([0.4113    , 0.37500985]),
 array([[&#39;1007&#39;, &#39;1007_1&#39;, &#39;2:&#39;, &#39;1:&#39;],
        [&#39;1007&#39;, &#39;1007_1&#39;, &#39;2:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_2&#39;, &#39;1:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_2&#39;, &#39;2:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_40&#39;, &#39;2:&#39;, &#39;1:&#39;],
        [&#39;1007&#39;, &#39;1007_40&#39;, &#39;0:&#39;, &#39;0:&#39;],
        [&#39;1007&#39;, &#39;1007_99&#39;, &#39;2:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_99&#39;, &#39;2:&#39;, &#39;1:&#39;],
        [&#39;1007&#39;, &#39;1007_5&#39;, &#39;2:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_5&#39;, &#39;2:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_3&#39;, &#39;1:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_3&#39;, &#39;2:&#39;, &#39;1:&#39;],
        [&#39;1007&#39;, &#39;1007_6&#39;, &#39;1:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_6&#39;, &#39;2:&#39;, &#39;1:&#39;],
        [&#39;1007&#39;, &#39;1007_4&#39;, &#39;1:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_4&#39;, &#39;2:&#39;, &#39;1:&#39;],
        [&#39;1007&#39;, &#39;1007_39&#39;, &#39;1:&#39;, &#39;2:&#39;],
        [&#39;1007&#39;, &#39;1007_39&#39;, &#39;2:&#39;, &#39;1:&#39;]], dtype=&#39;&lt;U7&#39;)]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test0.pickle&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
    <span class="n">genes1</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">genes1</span><span class="p">[</span><span class="s1">&#39;POTEH&#39;</span><span class="p">][</span><span class="s1">&#39;predata&#39;</span><span class="p">][</span><span class="s1">&#39;10J_128:0:0&#39;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[array([&#39;chr22:15690532:A:G&#39;], dtype=&#39;&lt;U18&#39;),
 array([0.0021]),
 array([[&#39;10J_128:0:0&#39;, &#39;10J_128_7&#39;, &#39;1:&#39;],
        [&#39;10J_128:0:0&#39;, &#39;10J_128_7&#39;, &#39;1:&#39;],
        [&#39;10J_128:0:0&#39;, &#39;10J_128_227&#39;, &#39;2:&#39;],
        [&#39;10J_128:0:0&#39;, &#39;10J_128_227&#39;, &#39;?:&#39;],
        [&#39;10J_128:0:0&#39;, &#39;10J_128_111&#39;, &#39;1:&#39;],
        [&#39;10J_128:0:0&#39;, &#39;10J_128_111&#39;, &#39;2:&#39;]], dtype=&#39;&lt;U11&#39;)]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tmp</span><span class="o">=</span><span class="n">update_haps_ped</span><span class="p">(</span><span class="n">genes1</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Functions-for-formating-the-input-of-linkage-analysis">Functions for formating the input of linkage analysis<a class="anchor-link" href="#Functions-for-formating-the-input-of-linkage-analysis"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_allele" class="doc_header"><code>get_allele</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L88" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_allele</code>(<strong><code>s</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="name_haps" class="doc_header"><code>name_haps</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L92" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>name_haps</code>(<strong><code>snps</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_fam_hap" class="doc_header"><code>get_fam_hap</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L98" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_fam_hap</code>(<strong><code>haps</code></strong>, <strong><code>variants</code></strong>, <strong><code>vcf</code></strong>=<em><code>None</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_fam_geno" class="doc_header"><code>get_fam_geno</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L127" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_fam_geno</code>(<strong><code>haps</code></strong>, <strong><code>variants</code></strong>, <strong><code>vcf</code></strong>=<em><code>None</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="format_haps_bunch" class="doc_header"><code>format_haps_bunch</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L156" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>format_haps_bunch</code>(<strong><code>dhaps</code></strong>, <strong><code>fam</code></strong>, <strong><code>vcfs</code></strong>=<em><code>None</code></em>, <strong><code>cutoff</code></strong>=<em><code>None</code></em>, <strong><code>haplotype</code></strong>=<em><code>True</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h4 id="test">test<a class="anchor-link" href="#test"> </a></h4>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Functions-of-heterogeneity">Functions of heterogeneity<a class="anchor-link" href="#Functions-of-heterogeneity"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="hlod_fun" class="doc_header"><code>hlod_fun</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L194" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>hlod_fun</code>(<strong><code>Li</code></strong>, <strong><code>sign</code></strong>=<em><code>1</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Functions-for-linkage-analysis">Functions for linkage analysis<a class="anchor-link" href="#Functions-for-linkage-analysis"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="calculate_ped_lod" class="doc_header"><code>calculate_ped_lod</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L200" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>calculate_ped_lod</code>(<strong><code>ped</code></strong>, <strong><code>afreq</code></strong>=<em><code>None</code></em>, <strong><code>rho</code></strong>=<em><code>0</code></em>, <strong><code>model</code></strong>=<em><code>'AD'</code></em>, <strong><code>chrom</code></strong>=<em><code>'AUTOSOMAL'</code></em>, <strong><code>penetrances</code></strong>=<em><code>[0.01, 0.9, 0.9]</code></em>, <strong><code>dfreq</code></strong>=<em><code>0.001</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="parallel_lods" class="doc_header"><code>parallel_lods</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L230" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>parallel_lods</code>(<strong><code>haps</code></strong>, <strong><code>afreqs</code></strong>=<em><code>None</code></em>, <strong><code>rho</code></strong>=<em><code>0</code></em>, <strong><code>model</code></strong>=<em><code>'AD'</code></em>, <strong><code>chrom</code></strong>=<em><code>'AUTOSOMAL'</code></em>, <strong><code>penetrances</code></strong>=<em><code>[0.01, 0.9, 0.9]</code></em>, <strong><code>dfreq</code></strong>=<em><code>0.001</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="linkage_analysis" class="doc_header"><code>linkage_analysis</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L241" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>linkage_analysis</code>()</p>
</blockquote>
<p>linkage analysis function</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h4 id="test">test<a class="anchor-link" href="#test"> </a></h4>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test0_AFcutoff0.0_linkage.lods&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="n">var_res</span><span class="o">=</span><span class="n">format_fam_lods</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span><span class="n">prefix</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>
<span class="n">var_sovs</span><span class="p">,</span><span class="n">best_sovs</span><span class="o">=</span><span class="p">[],[]</span>
<span class="k">for</span> <span class="n">var</span><span class="p">,</span><span class="n">res</span> <span class="ow">in</span> <span class="n">var_res</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">res</span><span class="o">=</span><span class="n">res</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">best_sov</span><span class="o">=</span><span class="p">[</span><span class="n">var</span><span class="p">,</span><span class="s1">&#39;LOD0.5&#39;</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">theta</span> <span class="ow">in</span> <span class="n">res</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">sov</span> <span class="o">=</span> <span class="n">minimize_scalar</span><span class="p">(</span><span class="n">hlod_fun</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">theta</span><span class="p">]),</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="n">bounds</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;bounded&#39;</span><span class="p">,</span> <span class="n">options</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;xatol&#39;</span><span class="p">:</span><span class="mf">1e-8</span><span class="p">})</span>
            <span class="n">var_sov</span><span class="o">=</span><span class="p">[</span><span class="n">var</span><span class="p">,</span><span class="n">theta</span><span class="p">,</span><span class="n">sov</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="o">-</span><span class="n">sov</span><span class="o">.</span><span class="n">fun</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">var_sov</span><span class="o">=</span><span class="p">[</span><span class="n">var</span><span class="p">,</span><span class="n">theta</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">var_sovs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">var_sov</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">best_sov</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">&lt;</span><span class="n">var_sov</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span> 
            <span class="n">best_sov</span><span class="o">=</span><span class="n">var_sov</span>
    <span class="n">best_sovs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">best_sov</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">)</span>
<span class="n">var_sovs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">var_sovs</span><span class="p">)</span>
<span class="n">best_sovs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">best_sovs</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>0.46859060414135456
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">sum_variant_lods</span><span class="p">(</span><span class="n">lods</span><span class="p">):</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">lod</span> <span class="ow">in</span> <span class="n">lods</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">m</span><span class="p">,</span><span class="n">l</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">lod</span><span class="p">[</span><span class="s1">&#39;MARKER&#39;</span><span class="p">],</span><span class="n">lod</span><span class="p">[</span><span class="s1">&#39;LOD&#39;</span><span class="p">]):</span>
            <span class="k">if</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">variants</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="o">+=</span> <span class="n">l</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">variants</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="o">=</span> <span class="n">l</span>
    <span class="n">var_lst</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">var</span><span class="p">,</span><span class="n">lod</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">snp</span> <span class="o">=</span> <span class="n">var</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span>
        <span class="n">var_lst</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">snp</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)</span><span class="o">+</span><span class="p">[</span><span class="n">snp</span><span class="p">,</span><span class="n">lod</span><span class="p">])</span>
    <span class="n">variants</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">var_lst</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span><span class="s1">&#39;POS&#39;</span><span class="p">,</span><span class="s1">&#39;A0&#39;</span><span class="p">,</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span><span class="s1">&#39;LOD&#39;</span><span class="p">])</span>
    <span class="n">variants</span><span class="o">.</span><span class="n">POS</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">POS</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">variants</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;POS&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">variants</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h5 id="test-linkage_analysis">test <a href="/SEQLinkage/Linkage.html#linkage_analysis"><code>linkage_analysis</code></a><a class="anchor-link" href="#test-linkage_analysis"> </a></h5>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">fam_vcf</span><span class="o">=</span><span class="s1">&#39;../data/wg20220520trimed/fam17_vcf_rmfounderwithoutvcf.pickle&#39;</span>
<span class="n">fam_path</span><span class="o">=</span><span class="s1">&#39;../data/new_trim_ped_rmfounderwithoutvcf.fam&#39;</span>
<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">fam_vcf</span><span class="p">):</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fam_vcf</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
        <span class="n">fam17_vcf</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="n">fam17</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">fam_path</span><span class="p">,</span><span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;fid&#39;</span><span class="p">,</span><span class="s1">&#39;iid&#39;</span><span class="p">,</span><span class="s1">&#39;fathid&#39;</span><span class="p">,</span><span class="s1">&#39;mothid&#39;</span><span class="p">,</span><span class="s1">&#39;sex&#39;</span><span class="p">,</span><span class="s1">&#39;ad&#39;</span><span class="p">])</span>
<span class="n">fam17</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">fam17</span><span class="o">.</span><span class="n">iid</span><span class="p">)</span>
<span class="n">fam17</span><span class="o">.</span><span class="n">ad</span><span class="p">[</span><span class="n">fam17</span><span class="o">.</span><span class="n">ad</span><span class="o">==-</span><span class="mi">9</span><span class="p">]</span><span class="o">=</span><span class="mi">0</span>
<span class="n">fam17_d</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">cutoff</span><span class="o">=</span><span class="mf">0.0</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">fam17</span><span class="o">.</span><span class="n">fid</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
    <span class="n">fam17_d</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">fam17</span><span class="p">[</span><span class="n">fam17</span><span class="o">.</span><span class="n">fid</span><span class="o">==</span><span class="n">i</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gene_genotype_file</span><span class="o">=</span><span class="s1">&#39;../data/wg20220725raretrimmed/chr22test/tmp/CACHE/chr22test0.pickle&#39;</span>
<span class="n">linkage_analysis</span><span class="p">(</span><span class="n">gene_genotype_file</span><span class="p">,</span><span class="n">fam17_d</span><span class="p">,</span><span class="n">fam17_vcf</span><span class="p">,</span><span class="n">cutoff</span><span class="p">,</span><span class="n">chp</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>exist! jump ../data/wg20220725raretrimmed/chr22test/tmp/CACHE/chr22test0_AFcutoff0.0_linkage.input
exist! jump ../data/wg20220725raretrimmed/chr22test/tmp/CACHE/chr22test0_AFcutoff0.0_linkage.lods
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>create ../data/wg20220725raretrimmed/chr22test/tmp/CACHE/chr22test0_AFcutoff0.0_linkage.besthlod
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gene_genotype_file</span><span class="o">=</span><span class="s1">&#39;../data/wg20220725raretrimmed/chr22test/tmp/CACHE/chr22test3.pickle&#39;</span>
<span class="n">linkage_analysis</span><span class="p">(</span><span class="n">gene_genotype_file</span><span class="p">,</span><span class="n">fam17_d</span><span class="p">,</span><span class="n">fam17_vcf</span><span class="p">,</span><span class="n">cutoff</span><span class="p">,</span><span class="n">chp</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>exist! jump ../data/wg20220725raretrimmed/chr22test/tmp/CACHE/chr22test3_AFcutoff0.0_linkage.input
exist! jump ../data/wg20220725raretrimmed/chr22test/tmp/CACHE/chr22test3_AFcutoff0.0_linkage.lods
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>create ../data/wg20220725raretrimmed/chr22test/tmp/CACHE/chr22test3_AFcutoff0.0_linkage.besthlod
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gene_genotype_file</span><span class="o">=</span><span class="s1">&#39;../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test0.pickle&#39;</span>
<span class="n">linkage_analysis</span><span class="p">(</span><span class="n">gene_genotype_file</span><span class="p">,</span><span class="n">fam17_d</span><span class="p">,</span><span class="n">fam17_vcf</span><span class="p">,</span><span class="n">cutoff</span><span class="p">,</span><span class="n">chp</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>exist! jump ../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test0_AFcutoff0.0_linkage.input
exist! jump ../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test0_AFcutoff0.0_linkage.lods
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>create ../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test0_AFcutoff0.0_linkage.besthlod
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gene_genotype_file</span><span class="o">=</span><span class="s1">&#39;../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test1.pickle&#39;</span>
<span class="n">linkage_analysis</span><span class="p">(</span><span class="n">gene_genotype_file</span><span class="p">,</span><span class="n">fam17_d</span><span class="p">,</span><span class="n">fam17_vcf</span><span class="p">,</span><span class="n">cutoff</span><span class="p">,</span><span class="n">chp</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>exist! jump ../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test1_AFcutoff0.0_linkage.input
exist! jump ../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test1_AFcutoff0.0_linkage.lods
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>create ../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/chr22test1_AFcutoff0.0_linkage.besthlod
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Functions-for-summarizing-results">Functions for summarizing results<a class="anchor-link" href="#Functions-for-summarizing-results"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="format_fam_lods" class="doc_header"><code>format_fam_lods</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L310" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>format_fam_lods</code>(<strong><code>res</code></strong>, <strong><code>cutoff</code></strong>=<em><code>0</code></em>, <strong><code>prefix</code></strong>=<em><code>False</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="rows2one" class="doc_header"><code>rows2one</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L332" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>rows2one</code>(<strong><code>x</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_lods_batch" class="doc_header"><code>get_lods_batch</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L338" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_lods_batch</code>(<strong><code>path</code></strong>, <strong><code>fams</code></strong>=<em><code>None</code></em>, <strong><code>phase</code></strong>=<em><code>False</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_lods_chrom" class="doc_header"><code>get_lods_chrom</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L358" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_lods_chrom</code>(<strong><code>prefix</code></strong>, <strong><code>fams</code></strong>=<em><code>None</code></em>, <strong><code>phase</code></strong>=<em><code>False</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_hlod_chrom" class="doc_header"><code>get_hlod_chrom</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L372" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_hlod_chrom</code>(<strong><code>prefix</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="summarize_lods" class="doc_header"><code>summarize_lods</code><a href="https://github.com/changebio/SEQLinkage/tree/master/SEQLinkage/Linkage.py#L389" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>summarize_lods</code>(<strong><code>input_lod</code></strong>, <strong><code>output_prefix</code></strong>, <strong><code>regions</code></strong>, <strong><code>fams</code></strong>=<em><code>None</code></em>, <strong><code>phase</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>two output files:one is lod scores from rho 0 to 0.5. another is lod at rho=0 and max lod score combined with chr, pos and name</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h4 id="test">test<a class="anchor-link" href="#test"> </a></h4>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">summarize_lods</span><span class="p">(</span><span class="s1">&#39;../data/wg20220725raretrimmed_phase/chr22test/tmp/CACHE/*_linkage.lods&#39;</span><span class="p">,</span><span class="s1">&#39;../data/wg20220725raretrimmed_phase/chr22test&#39;</span><span class="p">,</span><span class="n">regions</span><span class="p">,</span><span class="n">phase</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/tmp/2742267.1.high_mem.q/ipykernel_32628/1320168421.py:26: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  genes.LODmax[genes.LODmax&lt;0]=0
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">summarize_lods</span><span class="p">(</span><span class="s1">&#39;../data/wg20220725raretrimmed/chr22test/tmp/CACHE/*_linkage.lods&#39;</span><span class="p">,</span><span class="s1">&#39;../data/wg20220725raretrimmed/chr22test&#39;</span><span class="p">,</span><span class="n">regions</span><span class="p">,</span><span class="n">phase</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/tmp/2742267.1.high_mem.q/ipykernel_32628/1320168421.py:26: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  genes.LODmax[genes.LODmax&lt;0]=0
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

